{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>split</th>\n",
       "      <th>labels</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>TSS</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>fpkm_median</th>\n",
       "      <th>tpm_median</th>\n",
       "      <th>fpkm_uq_median</th>\n",
       "      <th>Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>train</td>\n",
       "      <td>0.128766</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>58347673</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>-4.563199</td>\n",
       "      <td>-3.172669</td>\n",
       "      <td>-4.445362</td>\n",
       "      <td>[3, 3, 1, 0, 0, 2, 0, 1, 1, 0, 2, 1, 1, 3, 2, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>train</td>\n",
       "      <td>-1.442332</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>50885633</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>-7.243318</td>\n",
       "      <td>-6.687800</td>\n",
       "      <td>-7.179188</td>\n",
       "      <td>[3, 2, 3, 1, 0, 0, 0, 1, 0, 3, 1, 3, 3, 3, 2, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>train</td>\n",
       "      <td>1.172033</td>\n",
       "      <td>A2M</td>\n",
       "      <td>9115927</td>\n",
       "      <td>chr12</td>\n",
       "      <td>-</td>\n",
       "      <td>5.214346</td>\n",
       "      <td>6.680884</td>\n",
       "      <td>5.391613</td>\n",
       "      <td>[0, 3, 3, 3, 3, 1, 1, 0, 3, 2, 0, 2, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000166535</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>train</td>\n",
       "      <td>-1.300226</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>8822627</td>\n",
       "      <td>chr12</td>\n",
       "      <td>+</td>\n",
       "      <td>-2.501443</td>\n",
       "      <td>-1.024150</td>\n",
       "      <td>-2.346665</td>\n",
       "      <td>[3, 2, 1, 3, 2, 2, 0, 0, 2, 1, 1, 0, 1, 1, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>train</td>\n",
       "      <td>0.157178</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>42720879</td>\n",
       "      <td>chr22</td>\n",
       "      <td>-</td>\n",
       "      <td>2.025064</td>\n",
       "      <td>3.517855</td>\n",
       "      <td>2.194465</td>\n",
       "      <td>[1, 0, 2, 2, 1, 0, 3, 2, 2, 3, 2, 2, 3, 2, 1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18049</th>\n",
       "      <td>ENSG00000250067</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>train</td>\n",
       "      <td>1.277756</td>\n",
       "      <td>YJEFN3</td>\n",
       "      <td>19639691</td>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>0.554245</td>\n",
       "      <td>2.024284</td>\n",
       "      <td>0.743644</td>\n",
       "      <td>[2, 0, 0, 0, 3, 2, 0, 2, 0, 2, 2, 0, 2, 2, 2, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18061</th>\n",
       "      <td>ENSG00000251655</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>train</td>\n",
       "      <td>-1.463212</td>\n",
       "      <td>PRB1</td>\n",
       "      <td>11508524</td>\n",
       "      <td>chr12</td>\n",
       "      <td>-</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>[0, 2, 1, 3, 0, 2, 0, 0, 0, 3, 0, 1, 1, 0, 3, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18071</th>\n",
       "      <td>ENSG00000254647</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>train</td>\n",
       "      <td>-1.463212</td>\n",
       "      <td>INS</td>\n",
       "      <td>2182434</td>\n",
       "      <td>chr11</td>\n",
       "      <td>-</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>[2, 1, 3, 3, 2, 2, 0, 1, 1, 1, 1, 0, 1, 3, 2, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18073</th>\n",
       "      <td>ENSG00000254685</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>train</td>\n",
       "      <td>0.350839</td>\n",
       "      <td>FPGT</td>\n",
       "      <td>74663926</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1.582171</td>\n",
       "      <td>3.110397</td>\n",
       "      <td>1.757621</td>\n",
       "      <td>[0, 0, 0, 1, 3, 2, 2, 1, 3, 0, 2, 1, 1, 0, 3, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18111</th>\n",
       "      <td>ENSG00000257591</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>train</td>\n",
       "      <td>-0.034731</td>\n",
       "      <td>ZNF625</td>\n",
       "      <td>12267529</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>-3.878321</td>\n",
       "      <td>-2.445362</td>\n",
       "      <td>-3.712173</td>\n",
       "      <td>[3, 2, 1, 0, 2, 2, 2, 2, 2, 1, 0, 2, 1, 0, 2, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15708 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id       gene_type  split    labels gene_name       TSS  \\\n",
       "0      ENSG00000121410  protein_coding  train  0.128766      A1BG  58347673   \n",
       "1      ENSG00000148584  protein_coding  train -1.442332      A1CF  50885633   \n",
       "2      ENSG00000175899  protein_coding  train  1.172033       A2M   9115927   \n",
       "3      ENSG00000166535  protein_coding  train -1.300226     A2ML1   8822627   \n",
       "4      ENSG00000128274  protein_coding  train  0.157178    A4GALT  42720879   \n",
       "...                ...             ...    ...       ...       ...       ...   \n",
       "18049  ENSG00000250067  protein_coding  train  1.277756    YJEFN3  19639691   \n",
       "18061  ENSG00000251655  protein_coding  train -1.463212      PRB1  11508524   \n",
       "18071  ENSG00000254647  protein_coding  train -1.463212       INS   2182434   \n",
       "18073  ENSG00000254685  protein_coding  train  0.350839      FPGT  74663926   \n",
       "18111  ENSG00000257591  protein_coding  train -0.034731    ZNF625  12267529   \n",
       "\n",
       "      chromosome_name strand  fpkm_median  tpm_median  fpkm_uq_median  \\\n",
       "0               chr19      -    -4.563199   -3.172669       -4.445362   \n",
       "1               chr10      -    -7.243318   -6.687800       -7.179188   \n",
       "2               chr12      -     5.214346    6.680884        5.391613   \n",
       "3               chr12      +    -2.501443   -1.024150       -2.346665   \n",
       "4               chr22      -     2.025064    3.517855        2.194465   \n",
       "...               ...    ...          ...         ...             ...   \n",
       "18049           chr19      +     0.554245    2.024284        0.743644   \n",
       "18061           chr12      -    -7.643856   -7.643856       -7.643856   \n",
       "18071           chr11      -    -7.643856   -7.643856       -7.643856   \n",
       "18073            chr1      +     1.582171    3.110397        1.757621   \n",
       "18111           chr19      -    -3.878321   -2.445362       -3.712173   \n",
       "\n",
       "                                                     Seq  \n",
       "0      [3, 3, 1, 0, 0, 2, 0, 1, 1, 0, 2, 1, 1, 3, 2, ...  \n",
       "1      [3, 2, 3, 1, 0, 0, 0, 1, 0, 3, 1, 3, 3, 3, 2, ...  \n",
       "2      [0, 3, 3, 3, 3, 1, 1, 0, 3, 2, 0, 2, 0, 0, 0, ...  \n",
       "3      [3, 2, 1, 3, 2, 2, 0, 0, 2, 1, 1, 0, 1, 1, 0, ...  \n",
       "4      [1, 0, 2, 2, 1, 0, 3, 2, 2, 3, 2, 2, 3, 2, 1, ...  \n",
       "...                                                  ...  \n",
       "18049  [2, 0, 0, 0, 3, 2, 0, 2, 0, 2, 2, 0, 2, 2, 2, ...  \n",
       "18061  [0, 2, 1, 3, 0, 2, 0, 0, 0, 3, 0, 1, 1, 0, 3, ...  \n",
       "18071  [2, 1, 3, 3, 2, 2, 0, 1, 1, 1, 1, 0, 1, 3, 2, ...  \n",
       "18073  [0, 0, 0, 1, 3, 2, 2, 1, 3, 0, 2, 1, 1, 0, 3, ...  \n",
       "18111  [3, 2, 1, 0, 2, 2, 2, 2, 2, 1, 0, 2, 1, 0, 2, ...  \n",
       "\n",
       "[15708 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df0 = pd.read_hdf('../dataset/Albert/df_alberto_0.h5', mode='r')\n",
    "df1 = pd.read_hdf('../dataset/Albert/df_alberto_1.h5', mode='r')\n",
    "df2 = pd.read_hdf('../dataset/Albert/df_alberto_2.h5', mode='r')\n",
    "df3 = pd.read_hdf('../dataset/Albert/df_alberto_3.h5', mode='r')\n",
    "df4 = pd.read_hdf('../dataset/Albert/df_alberto_4.h5', mode='r')\n",
    "df5 = pd.read_hdf('../dataset/Albert/df_alberto_5.h5', mode='r')\n",
    "df6 = pd.read_hdf('../dataset/Albert/df_alberto_6.h5', mode='r')\n",
    "\n",
    "df = pd.concat([df0, df1, df2, df3, df4, df5, df6])\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>Avg_Beta_Methylation</th>\n",
       "      <th>AddressA_ID</th>\n",
       "      <th>AlleleA_ProbeSeq</th>\n",
       "      <th>AddressB_ID</th>\n",
       "      <th>AlleleB_ProbeSeq</th>\n",
       "      <th>Infinium_Design_Type</th>\n",
       "      <th>Next_Base</th>\n",
       "      <th>Color_Channel</th>\n",
       "      <th>Forward_Sequence</th>\n",
       "      <th>Genome_Build</th>\n",
       "      <th>CHR</th>\n",
       "      <th>MAPINFO</th>\n",
       "      <th>SourceSeq</th>\n",
       "      <th>Chromosome_36</th>\n",
       "      <th>Coordinate_36</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Probe_SNPs</th>\n",
       "      <th>Probe_SNPs_10</th>\n",
       "      <th>Random_Loci</th>\n",
       "      <th>Methyl27_Loci</th>\n",
       "      <th>UCSC_RefGene_Name</th>\n",
       "      <th>UCSC_RefGene_Accession</th>\n",
       "      <th>UCSC_RefGene_Group</th>\n",
       "      <th>UCSC_CpG_Islands_Name</th>\n",
       "      <th>Relation_to_UCSC_CpG_Island</th>\n",
       "      <th>Phantom</th>\n",
       "      <th>DMR</th>\n",
       "      <th>Enhancer</th>\n",
       "      <th>HMM_Island</th>\n",
       "      <th>Regulatory_Feature_Name</th>\n",
       "      <th>Regulatory_Feature_Group</th>\n",
       "      <th>DHS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cg00000292</td>\n",
       "      <td>0.830354</td>\n",
       "      <td>43764508</td>\n",
       "      <td>AAAACATTAATTACCAACCRCTCTTCCAAAAAACACTTACCATTAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TGGGGTGAGTGAGACCACGGGCCTCACCCCGGACCAAGTTAAGCGG...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>16</td>\n",
       "      <td>28890100.0</td>\n",
       "      <td>CGGCCTCAATGGTAAGTGTCCCTTGGAAGAGCGGCTGGTAATTAAT...</td>\n",
       "      <td>16</td>\n",
       "      <td>28797601</td>\n",
       "      <td>+</td>\n",
       "      <td>rs62037371</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ATP2A1;ATP2A1</td>\n",
       "      <td>NM_004320;NM_173201</td>\n",
       "      <td>1stExon;1stExon</td>\n",
       "      <td>chr16:28890954-28891868</td>\n",
       "      <td>N_Shore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cg00002426</td>\n",
       "      <td>0.075181</td>\n",
       "      <td>48607471</td>\n",
       "      <td>CAATATAATAACATTACCTTACCCRTCTTATAATCAAACCAAACRA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCGCTGTCGACCAGCGCAGAATAATGCCACTTTTGATTGCAAAGTG...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>3</td>\n",
       "      <td>57743543.0</td>\n",
       "      <td>CGCTCTCGTCTGGTTTGATCACAAGACGGGCAAGGTAATGTCACCA...</td>\n",
       "      <td>3</td>\n",
       "      <td>57718583</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>SLMAP</td>\n",
       "      <td>NM_007159</td>\n",
       "      <td>1stExon</td>\n",
       "      <td>chr3:57741925-57742990</td>\n",
       "      <td>S_Shore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3:57718360-57718734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cg00003994</td>\n",
       "      <td>0.026089</td>\n",
       "      <td>22624445</td>\n",
       "      <td>TAATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGTGGTGGTGGTGGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGC...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>7</td>\n",
       "      <td>15725862.0</td>\n",
       "      <td>GGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGCAAACATGCCCTC...</td>\n",
       "      <td>7</td>\n",
       "      <td>15692387</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>MEOX2</td>\n",
       "      <td>NM_005924</td>\n",
       "      <td>1stExon</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDMR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7:15692387-15692746</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cg00005847</td>\n",
       "      <td>0.624374</td>\n",
       "      <td>27780345</td>\n",
       "      <td>ATACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACTTCAGTATGAAGTACCACCCGTTACATGAAAACCGATCATTGGC...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2</td>\n",
       "      <td>177029073.0</td>\n",
       "      <td>TACTGTAATGCACCCTGTATTTAAGGCACTGGGCTTGCCCCATTAA...</td>\n",
       "      <td>2</td>\n",
       "      <td>176737319</td>\n",
       "      <td>+</td>\n",
       "      <td>rs2857541</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>HOXD3</td>\n",
       "      <td>NM_006898</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>chr2:177029413-177029941</td>\n",
       "      <td>N_Shore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RDMR</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cg00006414</td>\n",
       "      <td>0.034145</td>\n",
       "      <td>26693387</td>\n",
       "      <td>ACTCRAAAACCRAACAAAACAAAACCCCAATACTAATCRTTAATAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TATTAAACAGACGGGGTCTGGCTGTTGCCCAGGCTGGAGGGCAGTG...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>7</td>\n",
       "      <td>148822837.0</td>\n",
       "      <td>CTCGGAAACCGAGCAGGGCAAAACCCCAGTGCTGATCGTTAGTGGG...</td>\n",
       "      <td>7</td>\n",
       "      <td>148453770</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ZNF425;ZNF398</td>\n",
       "      <td>NM_001001661;NM_020781</td>\n",
       "      <td>Body;TSS1500</td>\n",
       "      <td>chr7:148823194-148824228</td>\n",
       "      <td>N_Shore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7:148453721-148455161</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25973</th>\n",
       "      <td>cg27657283</td>\n",
       "      <td>0.020211</td>\n",
       "      <td>21628476</td>\n",
       "      <td>AAAACCAAAAATTCCTCTAACTACAACTTAAATACCCATAATCTTC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AATGACCTTCTGGGACCAAAGGTTCCTCTAACTGCAACTTGAGTGC...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1</td>\n",
       "      <td>43233523.0</td>\n",
       "      <td>GGACCAAAGGTTCCTCTAACTGCAACTTGAGTGCCCATAATCTTCC...</td>\n",
       "      <td>1</td>\n",
       "      <td>43006110</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>LEPRE1;C1orf50;LEPRE1</td>\n",
       "      <td>NM_022356;NM_024097;NM_001146289</td>\n",
       "      <td>TSS1500;Body;TSS1500</td>\n",
       "      <td>chr1:43232214-43233359</td>\n",
       "      <td>S_Shore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:43231970-43233594</td>\n",
       "      <td>Promoter_Associated</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25974</th>\n",
       "      <td>cg27661264</td>\n",
       "      <td>0.277928</td>\n",
       "      <td>19659318</td>\n",
       "      <td>CTAACTAATTTAAAATCTTTCTCCTCRCTTCTCCRCTAACCAAAAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTCCTGTACTCCAGCGTTCCGAGGCCCACGTCTTGCGTCACGTGTC...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>57427738.0</td>\n",
       "      <td>TGACTGATTTGGAGTCTTTCTCCTCGCTTCTCCGCTGGCCAGAGAA...</td>\n",
       "      <td>20</td>\n",
       "      <td>56861133</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>GNAS;GNAS;GNAS</td>\n",
       "      <td>NM_080425;NM_001077490;NM_016592</td>\n",
       "      <td>TSS1500;TSS1500;3'UTR</td>\n",
       "      <td>chr20:57427691-57427995</td>\n",
       "      <td>Island</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>20:56860807-56861390</td>\n",
       "      <td>20:57426788-57428499</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25975</th>\n",
       "      <td>cg27662379</td>\n",
       "      <td>0.011058</td>\n",
       "      <td>12769385</td>\n",
       "      <td>CCACTAAACAACCRTTCATTAAACAAAAACTAATAACACAAAATAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCGCCCTCCCGCCACTGAGCAGCCGTTCATTGGGCAAGGGCTGGTG...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>3</td>\n",
       "      <td>128369753.0</td>\n",
       "      <td>CGCCGCCATCTTGTGTCACCAGCCCTTGCCCAATGAACGGCTGCTC...</td>\n",
       "      <td>3</td>\n",
       "      <td>129852443</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>RPN1</td>\n",
       "      <td>NM_002950</td>\n",
       "      <td>TSS200</td>\n",
       "      <td>chr3:128368929-128369805</td>\n",
       "      <td>Island</td>\n",
       "      <td>high-CpG:129852099-129852456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3:129851620-129852544</td>\n",
       "      <td>3:128368829-128370060</td>\n",
       "      <td>Promoter_Associated</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25976</th>\n",
       "      <td>cg27662877</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>59703426</td>\n",
       "      <td>AACCTATACAAAATCCAATCTTATAAAAACRACRAACTAAACAACR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CGGGCATGGATGGCCTGTGCAGAGTCCAGTCTTGTGAGGGCGGCGA...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>18</td>\n",
       "      <td>71959595.0</td>\n",
       "      <td>GCCTGTGCAGAGTCCAGTCTTGTGAGGGCGGCGAACTAGGCAGCGA...</td>\n",
       "      <td>18</td>\n",
       "      <td>70110575</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>CYB5A;CYB5A</td>\n",
       "      <td>NM_148923;NM_001914</td>\n",
       "      <td>TSS1500;TSS1500</td>\n",
       "      <td>chr18:71958141-71959770</td>\n",
       "      <td>Island</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18:70109122-70110737</td>\n",
       "      <td>18:71958299-71959985</td>\n",
       "      <td>Promoter_Associated</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25977</th>\n",
       "      <td>cg27665659</td>\n",
       "      <td>0.014864</td>\n",
       "      <td>57766438</td>\n",
       "      <td>CRACCRAACTCCTAAAAACTACAATCAAAAACCTAATTACCATAAT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CGTTACCATGACGACCGGGCTCCTAGAAGCTGCAGTCAAGGACCTG...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>7</td>\n",
       "      <td>100797595.0</td>\n",
       "      <td>GACCGGGCTCCTAGAAGCTGCAGTCAAGGACCTGGTTGCCATGGTT...</td>\n",
       "      <td>7</td>\n",
       "      <td>100584315</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>AP1S1</td>\n",
       "      <td>NM_001283</td>\n",
       "      <td>TSS200</td>\n",
       "      <td>chr7:100797532-100798479</td>\n",
       "      <td>Island</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7:100584253-100585088</td>\n",
       "      <td>7:100797131-100798502</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25978 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            genes  Avg_Beta_Methylation  AddressA_ID  \\\n",
       "0      cg00000292              0.830354     43764508   \n",
       "1      cg00002426              0.075181     48607471   \n",
       "2      cg00003994              0.026089     22624445   \n",
       "3      cg00005847              0.624374     27780345   \n",
       "4      cg00006414              0.034145     26693387   \n",
       "...           ...                   ...          ...   \n",
       "25973  cg27657283              0.020211     21628476   \n",
       "25974  cg27661264              0.277928     19659318   \n",
       "25975  cg27662379              0.011058     12769385   \n",
       "25976  cg27662877              0.025641     59703426   \n",
       "25977  cg27665659              0.014864     57766438   \n",
       "\n",
       "                                        AlleleA_ProbeSeq  AddressB_ID  \\\n",
       "0      AAAACATTAATTACCAACCRCTCTTCCAAAAAACACTTACCATTAA...          NaN   \n",
       "1      CAATATAATAACATTACCTTACCCRTCTTATAATCAAACCAAACRA...          NaN   \n",
       "2      TAATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCT...          NaN   \n",
       "3      ATACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTA...          NaN   \n",
       "4      ACTCRAAAACCRAACAAAACAAAACCCCAATACTAATCRTTAATAA...          NaN   \n",
       "...                                                  ...          ...   \n",
       "25973  AAAACCAAAAATTCCTCTAACTACAACTTAAATACCCATAATCTTC...          NaN   \n",
       "25974  CTAACTAATTTAAAATCTTTCTCCTCRCTTCTCCRCTAACCAAAAA...          NaN   \n",
       "25975  CCACTAAACAACCRTTCATTAAACAAAAACTAATAACACAAAATAA...          NaN   \n",
       "25976  AACCTATACAAAATCCAATCTTATAAAAACRACRAACTAAACAACR...          NaN   \n",
       "25977  CRACCRAACTCCTAAAAACTACAATCAAAAACCTAATTACCATAAT...          NaN   \n",
       "\n",
       "      AlleleB_ProbeSeq Infinium_Design_Type Next_Base Color_Channel  \\\n",
       "0                  NaN                   II       NaN           NaN   \n",
       "1                  NaN                   II       NaN           NaN   \n",
       "2                  NaN                   II       NaN           NaN   \n",
       "3                  NaN                   II       NaN           NaN   \n",
       "4                  NaN                   II       NaN           NaN   \n",
       "...                ...                  ...       ...           ...   \n",
       "25973              NaN                   II       NaN           NaN   \n",
       "25974              NaN                   II       NaN           NaN   \n",
       "25975              NaN                   II       NaN           NaN   \n",
       "25976              NaN                   II       NaN           NaN   \n",
       "25977              NaN                   II       NaN           NaN   \n",
       "\n",
       "                                        Forward_Sequence  Genome_Build   CHR  \\\n",
       "0      TGGGGTGAGTGAGACCACGGGCCTCACCCCGGACCAAGTTAAGCGG...          37.0    16   \n",
       "1      CCGCTGTCGACCAGCGCAGAATAATGCCACTTTTGATTGCAAAGTG...          37.0     3   \n",
       "2      GGTGGTGGTGGTGGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGC...          37.0     7   \n",
       "3      ACTTCAGTATGAAGTACCACCCGTTACATGAAAACCGATCATTGGC...          37.0     2   \n",
       "4      TATTAAACAGACGGGGTCTGGCTGTTGCCCAGGCTGGAGGGCAGTG...          37.0     7   \n",
       "...                                                  ...           ...   ...   \n",
       "25973  AATGACCTTCTGGGACCAAAGGTTCCTCTAACTGCAACTTGAGTGC...          37.0     1   \n",
       "25974  CTCCTGTACTCCAGCGTTCCGAGGCCCACGTCTTGCGTCACGTGTC...          37.0  20.0   \n",
       "25975  CCGCCCTCCCGCCACTGAGCAGCCGTTCATTGGGCAAGGGCTGGTG...          37.0     3   \n",
       "25976  CGGGCATGGATGGCCTGTGCAGAGTCCAGTCTTGTGAGGGCGGCGA...          37.0    18   \n",
       "25977  CGTTACCATGACGACCGGGCTCCTAGAAGCTGCAGTCAAGGACCTG...          37.0     7   \n",
       "\n",
       "           MAPINFO                                          SourceSeq  \\\n",
       "0       28890100.0  CGGCCTCAATGGTAAGTGTCCCTTGGAAGAGCGGCTGGTAATTAAT...   \n",
       "1       57743543.0  CGCTCTCGTCTGGTTTGATCACAAGACGGGCAAGGTAATGTCACCA...   \n",
       "2       15725862.0  GGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGCAAACATGCCCTC...   \n",
       "3      177029073.0  TACTGTAATGCACCCTGTATTTAAGGCACTGGGCTTGCCCCATTAA...   \n",
       "4      148822837.0  CTCGGAAACCGAGCAGGGCAAAACCCCAGTGCTGATCGTTAGTGGG...   \n",
       "...            ...                                                ...   \n",
       "25973   43233523.0  GGACCAAAGGTTCCTCTAACTGCAACTTGAGTGCCCATAATCTTCC...   \n",
       "25974   57427738.0  TGACTGATTTGGAGTCTTTCTCCTCGCTTCTCCGCTGGCCAGAGAA...   \n",
       "25975  128369753.0  CGCCGCCATCTTGTGTCACCAGCCCTTGCCCAATGAACGGCTGCTC...   \n",
       "25976   71959595.0  GCCTGTGCAGAGTCCAGTCTTGTGAGGGCGGCGAACTAGGCAGCGA...   \n",
       "25977  100797595.0  GACCGGGCTCCTAGAAGCTGCAGTCAAGGACCTGGTTGCCATGGTT...   \n",
       "\n",
       "      Chromosome_36  Coordinate_36 Strand  Probe_SNPs Probe_SNPs_10  \\\n",
       "0                16       28797601      +  rs62037371           NaN   \n",
       "1                 3       57718583      +         NaN           NaN   \n",
       "2                 7       15692387      -         NaN           NaN   \n",
       "3                 2      176737319      +   rs2857541           NaN   \n",
       "4                 7      148453770      +         NaN           NaN   \n",
       "...             ...            ...    ...         ...           ...   \n",
       "25973             1       43006110      -         NaN           NaN   \n",
       "25974            20       56861133      +         NaN           NaN   \n",
       "25975             3      129852443      -         NaN           NaN   \n",
       "25976            18       70110575      -         NaN           NaN   \n",
       "25977             7      100584315      -         NaN           NaN   \n",
       "\n",
       "      Random_Loci  Methyl27_Loci      UCSC_RefGene_Name  \\\n",
       "0             NaN           True          ATP2A1;ATP2A1   \n",
       "1             NaN           True                  SLMAP   \n",
       "2             NaN           True                  MEOX2   \n",
       "3             NaN           True                  HOXD3   \n",
       "4             NaN           True          ZNF425;ZNF398   \n",
       "...           ...            ...                    ...   \n",
       "25973         NaN           True  LEPRE1;C1orf50;LEPRE1   \n",
       "25974         NaN           True         GNAS;GNAS;GNAS   \n",
       "25975         NaN           True                   RPN1   \n",
       "25976         NaN           True            CYB5A;CYB5A   \n",
       "25977         NaN           True                  AP1S1   \n",
       "\n",
       "                 UCSC_RefGene_Accession     UCSC_RefGene_Group  \\\n",
       "0                   NM_004320;NM_173201        1stExon;1stExon   \n",
       "1                             NM_007159                1stExon   \n",
       "2                             NM_005924                1stExon   \n",
       "3                             NM_006898                  5'UTR   \n",
       "4                NM_001001661;NM_020781           Body;TSS1500   \n",
       "...                                 ...                    ...   \n",
       "25973  NM_022356;NM_024097;NM_001146289   TSS1500;Body;TSS1500   \n",
       "25974  NM_080425;NM_001077490;NM_016592  TSS1500;TSS1500;3'UTR   \n",
       "25975                         NM_002950                 TSS200   \n",
       "25976               NM_148923;NM_001914        TSS1500;TSS1500   \n",
       "25977                         NM_001283                 TSS200   \n",
       "\n",
       "          UCSC_CpG_Islands_Name Relation_to_UCSC_CpG_Island  \\\n",
       "0       chr16:28890954-28891868                     N_Shore   \n",
       "1        chr3:57741925-57742990                     S_Shore   \n",
       "2                           NaN                         NaN   \n",
       "3      chr2:177029413-177029941                     N_Shore   \n",
       "4      chr7:148823194-148824228                     N_Shore   \n",
       "...                         ...                         ...   \n",
       "25973    chr1:43232214-43233359                     S_Shore   \n",
       "25974   chr20:57427691-57427995                      Island   \n",
       "25975  chr3:128368929-128369805                      Island   \n",
       "25976   chr18:71958141-71959770                      Island   \n",
       "25977  chr7:100797532-100798479                      Island   \n",
       "\n",
       "                            Phantom   DMR Enhancer             HMM_Island  \\\n",
       "0                               NaN   NaN      NaN                    NaN   \n",
       "1                               NaN   NaN      NaN    3:57718360-57718734   \n",
       "2                               NaN  CDMR      NaN    7:15692387-15692746   \n",
       "3                               NaN  RDMR     True                    NaN   \n",
       "4                               NaN   NaN      NaN  7:148453721-148455161   \n",
       "...                             ...   ...      ...                    ...   \n",
       "25973                           NaN   NaN      NaN                    NaN   \n",
       "25974                           NaN   NaN     True   20:56860807-56861390   \n",
       "25975  high-CpG:129852099-129852456   NaN      NaN  3:129851620-129852544   \n",
       "25976                           NaN   NaN      NaN   18:70109122-70110737   \n",
       "25977                           NaN   NaN      NaN  7:100584253-100585088   \n",
       "\n",
       "      Regulatory_Feature_Name Regulatory_Feature_Group   DHS  \n",
       "0                         NaN                      NaN   NaN  \n",
       "1                         NaN                      NaN   NaN  \n",
       "2                         NaN                      NaN   NaN  \n",
       "3                         NaN                      NaN   NaN  \n",
       "4                         NaN                      NaN   NaN  \n",
       "...                       ...                      ...   ...  \n",
       "25973     1:43231970-43233594      Promoter_Associated   NaN  \n",
       "25974    20:57426788-57428499             Unclassified  True  \n",
       "25975   3:128368829-128370060      Promoter_Associated   NaN  \n",
       "25976    18:71958299-71959985      Promoter_Associated   NaN  \n",
       "25977   7:100797131-100798502             Unclassified   NaN  \n",
       "\n",
       "[25978 rows x 33 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "meth= pd.read_csv('../dataset/methylation_chr.csv', sep=',')\n",
    "meth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'gene_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[14], line 38\u001b[0m\n\u001b[0;32m     35\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m count_satisfied, count_not_satisfied\n\u001b[0;32m     37\u001b[0m \u001b[38;5;66;03m# Raggruppa il dataframe 2 per gene_id, in modo da avere una lista di tutti i MAPINFO per ogni gene_id\u001b[39;00m\n\u001b[1;32m---> 38\u001b[0m df2_grouped \u001b[38;5;241m=\u001b[39m \u001b[43mdf2\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgroupby\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgene_id\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMAPINFO\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mapply(\u001b[38;5;28mlist\u001b[39m)\u001b[38;5;241m.\u001b[39mreset_index()\n\u001b[0;32m     40\u001b[0m \u001b[38;5;66;03m# Unisci il dataframe 1 con il dataframe raggruppato 2\u001b[39;00m\n\u001b[0;32m     41\u001b[0m merged_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mmerge(df1, df2_grouped, on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene_id\u001b[39m\u001b[38;5;124m'\u001b[39m, how\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mleft\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\dany0\\anaconda3\\envs\\AIE\\Lib\\site-packages\\pandas\\core\\frame.py:9170\u001b[0m, in \u001b[0;36mDataFrame.groupby\u001b[1;34m(self, by, axis, level, as_index, sort, group_keys, observed, dropna)\u001b[0m\n\u001b[0;32m   9167\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m level \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m by \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m   9168\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mYou have to supply one of \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mby\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m and \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlevel\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m-> 9170\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDataFrameGroupBy\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   9171\u001b[0m \u001b[43m    \u001b[49m\u001b[43mobj\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m   9172\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkeys\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mby\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   9173\u001b[0m \u001b[43m    \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   9174\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   9175\u001b[0m \u001b[43m    \u001b[49m\u001b[43mas_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mas_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   9176\u001b[0m \u001b[43m    \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   9177\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgroup_keys\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgroup_keys\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   9178\u001b[0m \u001b[43m    \u001b[49m\u001b[43mobserved\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mobserved\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   9179\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdropna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdropna\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   9180\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\dany0\\anaconda3\\envs\\AIE\\Lib\\site-packages\\pandas\\core\\groupby\\groupby.py:1329\u001b[0m, in \u001b[0;36mGroupBy.__init__\u001b[1;34m(self, obj, keys, axis, level, grouper, exclusions, selection, as_index, sort, group_keys, observed, dropna)\u001b[0m\n\u001b[0;32m   1326\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdropna \u001b[38;5;241m=\u001b[39m dropna\n\u001b[0;32m   1328\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m grouper \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m-> 1329\u001b[0m     grouper, exclusions, obj \u001b[38;5;241m=\u001b[39m \u001b[43mget_grouper\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   1330\u001b[0m \u001b[43m        \u001b[49m\u001b[43mobj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1331\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1332\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1333\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1334\u001b[0m \u001b[43m        \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1335\u001b[0m \u001b[43m        \u001b[49m\u001b[43mobserved\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mobserved\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mno_default\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mobserved\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1336\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdropna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdropna\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1337\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1339\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m observed \u001b[38;5;129;01mis\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mno_default:\n\u001b[0;32m   1340\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28many\u001b[39m(ping\u001b[38;5;241m.\u001b[39m_passed_categorical \u001b[38;5;28;01mfor\u001b[39;00m ping \u001b[38;5;129;01min\u001b[39;00m grouper\u001b[38;5;241m.\u001b[39mgroupings):\n",
      "File \u001b[1;32mc:\\Users\\dany0\\anaconda3\\envs\\AIE\\Lib\\site-packages\\pandas\\core\\groupby\\grouper.py:1043\u001b[0m, in \u001b[0;36mget_grouper\u001b[1;34m(obj, key, axis, level, sort, observed, validate, dropna)\u001b[0m\n\u001b[0;32m   1041\u001b[0m         in_axis, level, gpr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m, gpr, \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m   1042\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m-> 1043\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(gpr)\n\u001b[0;32m   1044\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(gpr, Grouper) \u001b[38;5;129;01mand\u001b[39;00m gpr\u001b[38;5;241m.\u001b[39mkey \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m   1045\u001b[0m     \u001b[38;5;66;03m# Add key to exclusions\u001b[39;00m\n\u001b[0;32m   1046\u001b[0m     exclusions\u001b[38;5;241m.\u001b[39madd(gpr\u001b[38;5;241m.\u001b[39mkey)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'gene_id'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Carica i due dataset in dataframe pandas\n",
    "df1 = df\n",
    "df2 = meth\n",
    "# Rimuovi il prefisso 'chr' dal dataframe 1\n",
    "df1['chromosome_name'] = df1['chromosome_name'].str.replace('chr', '')\n",
    "\n",
    "# Definisci la funzione per controllare le condizioni\n",
    "def check_conditions(gene_id, tss, chromosome_name, map_infos, strand):\n",
    "    # Seleziona i dati nel secondo dataset che corrispondono al gene_id, cromosoma e strand\n",
    "    subset_df = df2[(df2['CHR'] == chromosome_name) & (df2['Strand'] == strand)]\n",
    "    \n",
    "    # Calcola i valori di k\n",
    "    k = 2 ** 16\n",
    "    \n",
    "    # Conta il numero di mapinfo che soddisfano e non soddisfano le condizioni\n",
    "    count_satisfied = 0\n",
    "    count_not_satisfied = 0\n",
    "    \n",
    "    # Controlla se esiste almeno una riga nel secondo dataset che soddisfa le condizioni\n",
    "    for map_info in map_infos:\n",
    "        satisfied = False\n",
    "        for _, row in subset_df.iterrows():\n",
    "            if tss - k <= map_info <= tss + k:\n",
    "                if map_info % 5 == 1:  # Controlla se MAPINFO corrisponde a citosina\n",
    "                    satisfied = True\n",
    "                    break\n",
    "        if satisfied:\n",
    "            count_satisfied += 1\n",
    "        else:\n",
    "            count_not_satisfied += 1\n",
    "    \n",
    "    return count_satisfied, count_not_satisfied\n",
    "\n",
    "# Raggruppa il dataframe 2 per gene_id, in modo da avere una lista di tutti i MAPINFO per ogni gene_id\n",
    "df2_grouped = df2.groupby('gene_id')['MAPINFO'].apply(list).reset_index()\n",
    "\n",
    "# Unisci il dataframe 1 con il dataframe raggruppato 2\n",
    "merged_df = pd.merge(df1, df2_grouped, on='gene_id', how='left')\n",
    "\n",
    "# Applica la funzione check_conditions a ciascuna riga del dataframe\n",
    "merged_df[['num_satisfied', 'num_not_satisfied']] = merged_df.apply(lambda row: pd.Series(check_conditions(row['gene_id'], row['TSS'], row['chromosome_name'], row['MAPINFO'], row['strand'])), axis=1)\n",
    "\n",
    "# Stampare il risultato\n",
    "print(merged_df)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "AIE",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
