{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:8: SyntaxWarning: invalid escape sequence '\\G'\n",
      "<>:8: SyntaxWarning: invalid escape sequence '\\G'\n",
      "C:\\Users\\Danilo\\AppData\\Local\\Temp\\ipykernel_13416\\4216738839.py:8: SyntaxWarning: invalid escape sequence '\\G'\n",
      "  for i, record in enumerate(SeqIO.parse(\"..\\\\dataset\\\\\\GRCh37.primary_assembly.genome.fa\", \"fasta\")):\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "chrdict = {}\n",
    "\n",
    "for i, record in enumerate(SeqIO.parse(\"..\\\\dataset\\\\\\GRCh37.primary_assembly.genome.fa\", \"fasta\")):\n",
    "    chrdict[record.id] = record.seq\n",
    "\n",
    "\n",
    "# LE SE SEQUENZE DI BASI AZOTATE SONO MEMORIZZATE IN UN DIZIONARIO CHE USA COME CHIAVE \"chr#\" dove # indica\n",
    "# il numero del cromosoma oppure X/Y/M (M non ci interessa perché é il DNA mitocondriale che non viene usato per questo caso)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr10': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr11': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr12': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr13': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr14': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr15': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr16': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr17': Seq('AAGCTTCTCACCCTGTTCCTGCATAGATAATTGCATGACAATTGCCTTGTCCCT...GGT'),\n",
       " 'chr18': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr19': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr2': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr20': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr21': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr22': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr3': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr4': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr5': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr6': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr7': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr8': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr9': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chrM': Seq('GATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGG...ATG'),\n",
       " 'chrX': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chrY': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'GL000192.1': Seq('GAATTCATTCACCATTATTCTTTTATAATATTGCTATTTTATTATTCTTGATCA...TTC'),\n",
       " 'GL000225.1': Seq('GAATTCTCACATGTCTCTGCACTGATCACCCAGGTGATGTAACTCCTGTCTAGG...TTC'),\n",
       " 'GL000194.1': Seq('GATCCCTGCCCTAAAACTTTCCCCCCTCATGTCCAGCAAATGCTGCATGGAGCC...ATC'),\n",
       " 'GL000193.1': Seq('GAATTCATTTCTTTGATTTGTAGAGTGGCAGGGTACCTAAAGTTAACATTTGTC...TTC'),\n",
       " 'GL000200.1': Seq('TGGGTTGATAGGTTTTTAAGAATTCTTTTATGACTTCAAAAGTGTCACTTCGTT...CCT'),\n",
       " 'GL000222.1': Seq('GAATTCTATGTGAGGGACACACATTCAAAACCCAGCAGCAGTATTCTGGAATCC...TTC'),\n",
       " 'GL000212.1': Seq('GAATTCCTCCTTGCACATTTCTGCTTTATCTCATCTACACAGAAGTGATCCAAT...TTC'),\n",
       " 'GL000195.1': Seq('GAATTCCTCGTTCACACAGTTTCTTAAGCTTCCTGGGATGCGACCTGTGATGGC...TTC'),\n",
       " 'GL000223.1': Seq('GAATTCTCTGATGTCGAGTGAGGTTGGAGGCACGGCTAAAGGCCTTTCCACACT...TTC'),\n",
       " 'GL000224.1': Seq('GAATTCTGTTGATGATGGATGTGAGAGTGGTGAAGCCCTGCGGGATGATGTAGA...TTC'),\n",
       " 'GL000219.1': Seq('GAATTCTACCTGTGGGACAAACATTGAGACCTTTGTAACATTGTTCTGGAATGC...TTC'),\n",
       " 'GL000205.1': Seq('GAATTCTTGGTATTATTCCTTTTGAGTACCCCTATTTATGAGAGAGAAAACGAA...TTC'),\n",
       " 'GL000215.1': Seq('GAATTCCATAATATAAAGTTACTTTTTTAAAAAAAAATTAGACAATCAATTATC...TTC'),\n",
       " 'GL000216.1': Seq('GAATTCCATTGCATTCCATTCCATTTCATTCTATTACACTCCATTCCATTCCTT...TTC'),\n",
       " 'GL000217.1': Seq('GAATTCACTGGTCTTACCATATTCCCCATCTTCCTGAAGCAGCTGGATTGATAG...TTC'),\n",
       " 'GL000199.1': Seq('GAATTCTATATGAGGGAGAAAAACTCAGAACCTAGCAGCAGTGTTCTGGAATCC...TTC'),\n",
       " 'GL000211.1': Seq('GAATTCTTTTTCAAGTAAGTCAGGGATTTCTTCTTAGTTTGGATCCATTGCTGG...TTC'),\n",
       " 'GL000213.1': Seq('GAATTCAGTGAATGAGCCAAGATTCTCCAGAAACCGCTCAGCTTTGCAGTGATA...TTC'),\n",
       " 'GL000220.1': Seq('GAATTCAATACTTAAACATAAAACCTAATAAAATTCTTAGAAGTAAACATAAGG...TTC'),\n",
       " 'GL000218.1': Seq('GAATTCATGTAAAGAAAGACCATGGGTGTTTGTACTGGATTTAGTATTCATCAT...TTC'),\n",
       " 'GL000209.1': Seq('GATCAGCTTTCCCTTTGAAATCAGAGTACTAATCTGACTCATTGTTCCCTGAAT...ATC'),\n",
       " 'GL000221.1': Seq('GAATTCGTTGAGGAGCTTCTGGAAAGTGCACATTCTGACTCAGCAGGTATTGGA...TTC'),\n",
       " 'GL000214.1': Seq('GAATTCAGTGCAATTCAAAACAGTAATTTGTGGGAAAAATTAGATATGCAGATA...TTC'),\n",
       " 'GL000228.1': Seq('GAATTCTGTGCATTTTGTTTGATGGAATTTGGTATGTTTACCAAAATGGCAGAA...TTC'),\n",
       " 'GL000227.1': Seq('GATCCAACTATATTGTCCACAAGATACTCACTTTAGATCAAAATACACAATGAG...ATC'),\n",
       " 'GL000191.1': Seq('GATCCACCTGCCTCAGCCTCCCAGAGTGCTGGGATTATAGGTGTGAGCCACTGC...ATC'),\n",
       " 'GL000208.1': Seq('GAATTCTTCAAAGAGTTCCAGATATCCACAGGCAGATTCTACAAAAGAAGTGTT...GAT'),\n",
       " 'GL000198.1': Seq('GAATTCGTCTCATAGAGTTGAACCTTTCTTTTCATTGATCAGTTTGGAAACACT...TTC'),\n",
       " 'GL000204.1': Seq('GATCAGGAAGAGTCCTGGGGGGCGGGGGGCGAGCACAGGCAGTGGAAGGGAAGC...ATC'),\n",
       " 'GL000233.1': Seq('GATCCCATTTGTCAATTTTGGATTTTGTGGCCATTGCTTCTGTTGTTTTAGACA...ATC'),\n",
       " 'GL000237.1': Seq('GATCTCAGGAGGCGGAGGTTGCAGTGATCGGTGGCCATAGAAGCCCTCCTGAGC...ATC'),\n",
       " 'GL000230.1': Seq('GATCCTTTGAGCCCAGGAGTTTGAGACCACCCTGGGAAACATGGCAAAACCCTC...ATC'),\n",
       " 'GL000242.1': Seq('GATCATGCCTGGGACACACAGCAAGACCTTGTCCCTTAAAAAAAAAAAAAAAAA...ATC'),\n",
       " 'GL000243.1': Seq('GATCTTCAGGCTCTGTCTCTGATTCTAGTGCTCTTGTTATTTCCACCATATCTG...ATC'),\n",
       " 'GL000241.1': Seq('GATCTGAAGACAAATCAGGGTTGCAGGATGGTTAAATATCTGCAAGTCAATAAA...ATC'),\n",
       " 'GL000236.1': Seq('GATCACATGGACACAGGAAGGGGAATATCACACTCTGGGGACTGTTGTGGGGTG...ATC'),\n",
       " 'GL000240.1': Seq('GATCAGTGGAGACCTGGTCAGCTGGGGCTTAGTGCTGGCCTGGTCAGCATGGGC...ATC'),\n",
       " 'GL000206.1': Seq('GATCCCACCAGAGGCCATAAACAGATAGTTCAGAGCTATGAGAACCCAGGTGGC...ATC'),\n",
       " 'GL000232.1': Seq('GATCAGGCCCAGGAACTGCTGCCTTGGGTAGCTTTGGAACAAGGCTCCTGCATC...ATC'),\n",
       " 'GL000234.1': Seq('GATCTCAATGCATCCAGATTCATAATATGTTTGCCTTTGAATGGAATGATCCAG...ATC'),\n",
       " 'GL000202.1': Seq('TGCCAGAAAAGGGTGTGTGTTGGAGGGGGTGTCTTTCACAATATATTTTTAATT...AGC'),\n",
       " 'GL000238.1': Seq('GATCACAGTATCACAATTGACCTTAAAACAATAAATTTTATAATCAGTGTAAAT...ATC'),\n",
       " 'GL000244.1': Seq('GATCTGCGCACCTTGGCCTCCCAAAGTGCTGGGATTACAGGCATGAGCCACCGT...ATC'),\n",
       " 'GL000248.1': Seq('GATCTGTCATTGTCTCCCATGCATGCTTCCCAGCATGCCTGCCTAGCCCCATTC...ATC'),\n",
       " 'GL000196.1': Seq('GATCCCACCAGAGTAGAGGTGGCTATACTAACAACAGAGAAAACAGTCTAAATC...ATC'),\n",
       " 'GL000249.1': Seq('GATCACCAAGGCTGGGAAATATTAGGGAAGGGGGGTTGCCAGACTCCGGACCTG...ATC'),\n",
       " 'GL000246.1': Seq('GATCTTAAGCCTTTGCTGGATGCATCGTTTGCAAATATCTTCTCCCATTGTGTA...ATC'),\n",
       " 'GL000203.1': Seq('GATCCCACAGGATGTGTCTATCTTAAGATTTAGATAAATATTTAAATTTTTTTC...ATC'),\n",
       " 'GL000197.1': Seq('CAAGACTCCTCAGAGTAAGGCATCTGTCCTATTTTGTTCCCCACCCAGGAGTTG...TGG'),\n",
       " 'GL000245.1': Seq('GATCAGATAGGCTTTTGGCTCATTGTAATGAGATGTTCCTATAAACCTAATTTG...ATC'),\n",
       " 'GL000247.1': Seq('GATCTAAGTTTGATTTTTCTGTACAACTCCTTTTCTGAAGTTTTACTCATTTCC...ATC'),\n",
       " 'GL000201.1': Seq('GATCCCACCCAAAGCAGCAGCTTTTTTTTTTTTTTTTTTTTTTTTTGTTATTTT...ATC'),\n",
       " 'GL000235.1': Seq('GATCTCTGAGGTGATGCAACTCTTGTCTAGGCACTGCCTACAGGGGACATTGGT...ATC'),\n",
       " 'GL000239.1': Seq('GATCTCTCTATCGGCCTGCTCTGGGCCACAGAGACTGGCCCTGACGACTTCTCA...ATC'),\n",
       " 'GL000210.1': Seq('AAAAGAACAGCTGGCCTGGGCTTGCTGCTGGGATGCGGGTGGGGCTCTGGGGGG...TCA'),\n",
       " 'GL000231.1': Seq('TTTAATGAATAAGTATTTACGCTATTAATATTTATAATATTGGTGCTATTATTT...ATC'),\n",
       " 'GL000229.1': Seq('GCCGCTGTGGGATTTTCCGTGGGAGAGGTATGGGAGAGGACTGACGCGGCTTCC...ATC'),\n",
       " 'GL000226.1': Seq('GATCTTTCTTTTGATAGAGCAGTTCTGAAAAACACTTTTTGTTGAATCTGCAAG...ATC'),\n",
       " 'GL000207.1': Seq('CATACATTTAATATACCCTCACCATACAGAATGTTCTTTCCCTATTACATAAGG...ATC')}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>p</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100008587</td>\n",
       "      <td>100008589</td>\n",
       "      <td>+</td>\n",
       "      <td>p1</td>\n",
       "      <td>CU680531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>-</td>\n",
       "      <td>p2</td>\n",
       "      <td>LOXL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>-</td>\n",
       "      <td>p3</td>\n",
       "      <td>LOXL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>-</td>\n",
       "      <td>p1</td>\n",
       "      <td>LOXL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>-</td>\n",
       "      <td>p1</td>\n",
       "      <td>PYROXD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048119</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9958814</td>\n",
       "      <td>9958815</td>\n",
       "      <td>-</td>\n",
       "      <td>p</td>\n",
       "      <td>chrY:9958814..9958815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048120</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9959226</td>\n",
       "      <td>9959229</td>\n",
       "      <td>-</td>\n",
       "      <td>p</td>\n",
       "      <td>chrY:9959226..9959229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048121</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9968102</td>\n",
       "      <td>9968103</td>\n",
       "      <td>-</td>\n",
       "      <td>p</td>\n",
       "      <td>chrY:9968102..9968103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048122</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9985705</td>\n",
       "      <td>9985711</td>\n",
       "      <td>-</td>\n",
       "      <td>p</td>\n",
       "      <td>chrY:9985705..9985711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048123</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9987854</td>\n",
       "      <td>9987855</td>\n",
       "      <td>-</td>\n",
       "      <td>p</td>\n",
       "      <td>chrY:9987854..9987855</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1048124 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome_name      start        end strand   p  \\\n",
       "0                 chr10  100008587  100008589      +  p1   \n",
       "1                 chr10  100015362  100015397      -  p2   \n",
       "2                 chr10  100017518  100017519      -  p3   \n",
       "3                 chr10  100027943  100027958      -  p1   \n",
       "4                 chr10  100174900  100174956      -  p1   \n",
       "...                 ...        ...        ...    ...  ..   \n",
       "1048119            chrY    9958814    9958815      -   p   \n",
       "1048120            chrY    9959226    9959229      -   p   \n",
       "1048121            chrY    9968102    9968103      -   p   \n",
       "1048122            chrY    9985705    9985711      -   p   \n",
       "1048123            chrY    9987854    9987855      -   p   \n",
       "\n",
       "                     gene_name  \n",
       "0                     CU680531  \n",
       "1                        LOXL4  \n",
       "2                        LOXL4  \n",
       "3                        LOXL4  \n",
       "4                      PYROXD2  \n",
       "...                        ...  \n",
       "1048119  chrY:9958814..9958815  \n",
       "1048120  chrY:9959226..9959229  \n",
       "1048121  chrY:9968102..9968103  \n",
       "1048122  chrY:9985705..9985711  \n",
       "1048123  chrY:9987854..9987855  \n",
       "\n",
       "[1048124 rows x 6 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# UA INIZIA INVECE LA PARTE DI LETTURA DEL FILE CONTENENTE I TSS\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "def extract_p(p):\n",
    "    return p.split(\"@\")[0]\n",
    "\n",
    "def extract_gene_name(string):\n",
    "    string =  string.split(\"@\")[1]\n",
    "    return string.split(\",\")[0]\n",
    "\n",
    "v1 = np.vectorize(extract_p)\n",
    "v2 = np.vectorize(extract_gene_name)\n",
    "\n",
    "df = pd.read_csv(\"..\\\\dataset\\\\TSS_human.bed\", delimiter=\"\\t\", skiprows=1, names=[\"chromosome_name\", \"start\", \"end\", \"locus_id\", \"color\", \"strand\", \"start2\", \"end2\", \"RGB\"])\n",
    "\n",
    "\n",
    "\n",
    "# Applica la funzione alla colonna del DataFrame\n",
    "df['p'] = v1(df['locus_id'].values)\n",
    "df['gene_name'] = v2(df['locus_id'].values)\n",
    "df = df.drop(columns=[\"locus_id\", \"color\", \"start2\", \"end2\", \"RGB\"])\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>p</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100008587</td>\n",
       "      <td>100008589</td>\n",
       "      <td>+</td>\n",
       "      <td>p1</td>\n",
       "      <td>CU680531</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>-</td>\n",
       "      <td>p2</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>-</td>\n",
       "      <td>p3</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>-</td>\n",
       "      <td>p1</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>-</td>\n",
       "      <td>p1</td>\n",
       "      <td>PYROXD2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294760</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7142165</td>\n",
       "      <td>7142189</td>\n",
       "      <td>+</td>\n",
       "      <td>p3</td>\n",
       "      <td>PRKY</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294761</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7142206</td>\n",
       "      <td>7142215</td>\n",
       "      <td>+</td>\n",
       "      <td>p9</td>\n",
       "      <td>PRKY</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294762</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7200834</td>\n",
       "      <td>7200838</td>\n",
       "      <td>+</td>\n",
       "      <td>p12</td>\n",
       "      <td>PRKY</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294763</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7200913</td>\n",
       "      <td>7200914</td>\n",
       "      <td>+</td>\n",
       "      <td>p14</td>\n",
       "      <td>PRKY</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294764</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9928241</td>\n",
       "      <td>9928245</td>\n",
       "      <td>+</td>\n",
       "      <td>p1</td>\n",
       "      <td>ENST00000416843</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>294765 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome_name      start        end strand    p        gene_name  \\\n",
       "0                chr10  100008587  100008589      +   p1         CU680531   \n",
       "1                chr10  100015362  100015397      -   p2            LOXL4   \n",
       "2                chr10  100017518  100017519      -   p3            LOXL4   \n",
       "3                chr10  100027943  100027958      -   p1            LOXL4   \n",
       "4                chr10  100174900  100174956      -   p1          PYROXD2   \n",
       "...                ...        ...        ...    ...  ...              ...   \n",
       "294760            chrY    7142165    7142189      +   p3             PRKY   \n",
       "294761            chrY    7142206    7142215      +   p9             PRKY   \n",
       "294762            chrY    7200834    7200838      +  p12             PRKY   \n",
       "294763            chrY    7200913    7200914      +  p14             PRKY   \n",
       "294764            chrY    9928241    9928245      +   p1  ENST00000416843   \n",
       "\n",
       "        p_value  \n",
       "0           1.0  \n",
       "1           2.0  \n",
       "2           3.0  \n",
       "3           1.0  \n",
       "4           1.0  \n",
       "...         ...  \n",
       "294760      3.0  \n",
       "294761      9.0  \n",
       "294762     12.0  \n",
       "294763     14.0  \n",
       "294764      1.0  \n",
       "\n",
       "[294765 rows x 7 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "def estrai_numero(s):\n",
    "    match = re.match(r'p(\\d+)', s)\n",
    "    if match:\n",
    "        return int(match.group(1))\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "df['p_value'] = df['p'].apply(estrai_numero)\n",
    "#droppa righe con p_value non definito\n",
    "df = df.dropna()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>65039</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58858938</td>\n",
       "      <td>58859039</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65040</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58859101</td>\n",
       "      <td>58859149</td>\n",
       "      <td>+</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4033</th>\n",
       "      <td>chr10</td>\n",
       "      <td>52645379</td>\n",
       "      <td>52645393</td>\n",
       "      <td>-</td>\n",
       "      <td>A1CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27660</th>\n",
       "      <td>chr13</td>\n",
       "      <td>101240987</td>\n",
       "      <td>101241049</td>\n",
       "      <td>-</td>\n",
       "      <td>A2LD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27124</th>\n",
       "      <td>chr12</td>\n",
       "      <td>9268507</td>\n",
       "      <td>9268523</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287241</th>\n",
       "      <td>chr9</td>\n",
       "      <td>131012693</td>\n",
       "      <td>131012727</td>\n",
       "      <td>-</td>\n",
       "      <td>uc011mav.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159387</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135895676</td>\n",
       "      <td>135895723</td>\n",
       "      <td>+</td>\n",
       "      <td>uc011mct.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288056</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138899746</td>\n",
       "      <td>138899753</td>\n",
       "      <td>-</td>\n",
       "      <td>uc011mds.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290683</th>\n",
       "      <td>chrM</td>\n",
       "      <td>6330</td>\n",
       "      <td>6392</td>\n",
       "      <td>+</td>\n",
       "      <td>uc011mfh.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290713</th>\n",
       "      <td>chrM</td>\n",
       "      <td>8423</td>\n",
       "      <td>8463</td>\n",
       "      <td>+</td>\n",
       "      <td>uc011mfi.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39802 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome_name      start        end strand   gene_name\n",
       "65039            chr19   58858938   58859039      -        A1BG\n",
       "65040            chr19   58859101   58859149      +    A1BG-AS1\n",
       "4033             chr10   52645379   52645393      -        A1CF\n",
       "27660            chr13  101240987  101241049      -       A2LD1\n",
       "27124            chr12    9268507    9268523      -         A2M\n",
       "...                ...        ...        ...    ...         ...\n",
       "287241            chr9  131012693  131012727      -  uc011mav.1\n",
       "159387            chr9  135895676  135895723      +  uc011mct.1\n",
       "288056            chr9  138899746  138899753      -  uc011mds.1\n",
       "290683            chrM       6330       6392      +  uc011mfh.1\n",
       "290713            chrM       8423       8463      +  uc011mfi.1\n",
       "\n",
       "[39802 rows x 5 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.loc[df.groupby(\"gene_name\")[\"p_value\"].idxmin()]\n",
    "df.drop(columns=[\"p\", \"p_value\"], inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chromosome_name\n",
       "chr1     3852\n",
       "chr2     2840\n",
       "chr11    2230\n",
       "chr3     2212\n",
       "chr12    2174\n",
       "chr19    2142\n",
       "chr7     2137\n",
       "chr6     2121\n",
       "chr17    2091\n",
       "chr5     2015\n",
       "chr14    1772\n",
       "chr4     1683\n",
       "chr8     1621\n",
       "chr10    1562\n",
       "chr9     1488\n",
       "chrX     1374\n",
       "chr16    1358\n",
       "chr15    1326\n",
       "chr20    1045\n",
       "chr22     946\n",
       "chr13     748\n",
       "chr18     512\n",
       "chr21     463\n",
       "chrY       53\n",
       "chrM       37\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['chromosome_name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_stable_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>TSS</th>\n",
       "      <th>strand</th>\n",
       "      <th>chromosome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>ENSG00000121410.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>58347673</td>\n",
       "      <td>-</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>ENSG00000148584.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50885633</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>ENSG00000175899.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>9115927</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "      <td>ENSG00000166535.20</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>8822627</td>\n",
       "      <td>+</td>\n",
       "      <td>chr12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>ENSG00000128274.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>42720879</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15615</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>ENSG00000203995.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>52842770</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15616</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>ENSG00000162378.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>52726478</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15617</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>ENSG00000159840.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>143381361</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15618</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>ENSG00000074755.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4143055</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15619</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>77683457</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15620 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name   gene_stable_id             gene_id       gene_type  \\\n",
       "0          A1BG  ENSG00000121410  ENSG00000121410.12  protein_coding   \n",
       "1          A1CF  ENSG00000148584  ENSG00000148584.15  protein_coding   \n",
       "2           A2M  ENSG00000175899  ENSG00000175899.15  protein_coding   \n",
       "3         A2ML1  ENSG00000166535  ENSG00000166535.20  protein_coding   \n",
       "4        A4GALT  ENSG00000128274  ENSG00000128274.17  protein_coding   \n",
       "...         ...              ...                 ...             ...   \n",
       "15615    ZYG11A  ENSG00000203995  ENSG00000203995.10  protein_coding   \n",
       "15616    ZYG11B  ENSG00000162378  ENSG00000162378.13  protein_coding   \n",
       "15617       ZYX  ENSG00000159840  ENSG00000159840.16  protein_coding   \n",
       "15618     ZZEF1  ENSG00000074755  ENSG00000074755.15  protein_coding   \n",
       "15619      ZZZ3  ENSG00000036549  ENSG00000036549.13  protein_coding   \n",
       "\n",
       "             TSS strand chromosome_name  \n",
       "0       58347673      -           chr19  \n",
       "1       50885633      -           chr10  \n",
       "2        9115927      -           chr12  \n",
       "3        8822627      +           chr12  \n",
       "4       42720879      -           chr22  \n",
       "...          ...    ...             ...  \n",
       "15615   52842770      +            chr1  \n",
       "15616   52726478      +            chr1  \n",
       "15617  143381361      +            chr7  \n",
       "15618    4143055      -           chr17  \n",
       "15619   77683457      -            chr1  \n",
       "\n",
       "[15620 rows x 7 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lizio=pd.read_csv(\"..\\\\dataset\\\\Join_with_Lizio.csv\",sep=\",\")\n",
    "lizio.drop(columns=[\"Unnamed: 0\"],inplace=True)\n",
    "lizio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>TSS_x</th>\n",
       "      <th>genomic_strand</th>\n",
       "      <th>chromosome_name_x</th>\n",
       "      <th>sequence</th>\n",
       "      <th>split</th>\n",
       "      <th>labels</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>gene_stable_id</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>TSS_y</th>\n",
       "      <th>strand</th>\n",
       "      <th>chromosome_name_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>58347673</td>\n",
       "      <td>-</td>\n",
       "      <td>chr19</td>\n",
       "      <td>[0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 2, ...</td>\n",
       "      <td>train</td>\n",
       "      <td>0.128766</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>58347673</td>\n",
       "      <td>-</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50885633</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>[1, 1, 3, 0, 0, 0, 3, 2, 2, 1, 1, 1, 2, 1, 2, ...</td>\n",
       "      <td>train</td>\n",
       "      <td>-1.442332</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50885633</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>9115927</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12</td>\n",
       "      <td>[1, 2, 0, 3, 1, 1, 0, 1, 2, 1, 2, 2, 1, 1, 0, ...</td>\n",
       "      <td>train</td>\n",
       "      <td>1.172033</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>9115927</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>ENSG00000166535.20</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>8822627</td>\n",
       "      <td>+</td>\n",
       "      <td>chr12</td>\n",
       "      <td>[3, 1, 3, 0, 3, 2, 2, 0, 2, 2, 3, 1, 3, 2, 2, ...</td>\n",
       "      <td>train</td>\n",
       "      <td>-1.300226</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>8822627</td>\n",
       "      <td>+</td>\n",
       "      <td>chr12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>42720879</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22</td>\n",
       "      <td>[1, 0, 1, 1, 2, 0, 3, 1, 0, 1, 3, 0, 1, 3, 0, ...</td>\n",
       "      <td>train</td>\n",
       "      <td>0.157178</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>42720879</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15615</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ENSG00000203995.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>52842770</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[3, 3, 2, 1, 3, 0, 3, 3, 2, 0, 1, 3, 0, 2, 0, ...</td>\n",
       "      <td>train</td>\n",
       "      <td>-0.984208</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>52842770</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15616</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>52726478</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[1, 3, 2, 2, 0, 3, 0, 0, 1, 2, 1, 1, 3, 3, 2, ...</td>\n",
       "      <td>train</td>\n",
       "      <td>0.694430</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>52726478</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15617</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>143381361</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7</td>\n",
       "      <td>[0, 0, 1, 3, 1, 3, 2, 2, 0, 3, 1, 3, 1, 1, 1, ...</td>\n",
       "      <td>train</td>\n",
       "      <td>1.646464</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>143381361</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15618</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4143055</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>[2, 1, 0, 2, 2, 1, 1, 1, 1, 3, 2, 0, 0, 0, 1, ...</td>\n",
       "      <td>train</td>\n",
       "      <td>0.765210</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4143055</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15619</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>77683457</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[0, 1, 0, 2, 0, 1, 3, 1, 0, 0, 1, 3, 2, 1, 0, ...</td>\n",
       "      <td>train</td>\n",
       "      <td>0.672941</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>77683457</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15620 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name_x             gene_id     gene_type_x      TSS_x  \\\n",
       "0            A1BG  ENSG00000121410.12  protein_coding   58347673   \n",
       "1            A1CF  ENSG00000148584.15  protein_coding   50885633   \n",
       "2             A2M  ENSG00000175899.15  protein_coding    9115927   \n",
       "3           A2ML1  ENSG00000166535.20  protein_coding    8822627   \n",
       "4          A4GALT  ENSG00000128274.17  protein_coding   42720879   \n",
       "...           ...                 ...             ...        ...   \n",
       "15615      ZYG11A  ENSG00000203995.10  protein_coding   52842770   \n",
       "15616      ZYG11B  ENSG00000162378.13  protein_coding   52726478   \n",
       "15617         ZYX  ENSG00000159840.16  protein_coding  143381361   \n",
       "15618       ZZEF1  ENSG00000074755.15  protein_coding    4143055   \n",
       "15619        ZZZ3  ENSG00000036549.13  protein_coding   77683457   \n",
       "\n",
       "      genomic_strand chromosome_name_x  \\\n",
       "0                  -             chr19   \n",
       "1                  -             chr10   \n",
       "2                  -             chr12   \n",
       "3                  +             chr12   \n",
       "4                  -             chr22   \n",
       "...              ...               ...   \n",
       "15615              +              chr1   \n",
       "15616              +              chr1   \n",
       "15617              +              chr7   \n",
       "15618              -             chr17   \n",
       "15619              -              chr1   \n",
       "\n",
       "                                                sequence  split    labels  \\\n",
       "0      [0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 2, ...  train  0.128766   \n",
       "1      [1, 1, 3, 0, 0, 0, 3, 2, 2, 1, 1, 1, 2, 1, 2, ...  train -1.442332   \n",
       "2      [1, 2, 0, 3, 1, 1, 0, 1, 2, 1, 2, 2, 1, 1, 0, ...  train  1.172033   \n",
       "3      [3, 1, 3, 0, 3, 2, 2, 0, 2, 2, 3, 1, 3, 2, 2, ...  train -1.300226   \n",
       "4      [1, 0, 1, 1, 2, 0, 3, 1, 0, 1, 3, 0, 1, 3, 0, ...  train  0.157178   \n",
       "...                                                  ...    ...       ...   \n",
       "15615  [3, 3, 2, 1, 3, 0, 3, 3, 2, 0, 1, 3, 0, 2, 0, ...  train -0.984208   \n",
       "15616  [1, 3, 2, 2, 0, 3, 0, 0, 1, 2, 1, 1, 3, 3, 2, ...  train  0.694430   \n",
       "15617  [0, 0, 1, 3, 1, 3, 2, 2, 0, 3, 1, 3, 1, 1, 1, ...  train  1.646464   \n",
       "15618  [2, 1, 0, 2, 2, 1, 1, 1, 1, 3, 2, 0, 0, 0, 1, ...  train  0.765210   \n",
       "15619  [0, 1, 0, 2, 0, 1, 3, 1, 0, 0, 1, 3, 2, 1, 0, ...  train  0.672941   \n",
       "\n",
       "      gene_name_y   gene_stable_id     gene_type_y      TSS_y strand  \\\n",
       "0            A1BG  ENSG00000121410  protein_coding   58347673      -   \n",
       "1            A1CF  ENSG00000148584  protein_coding   50885633      -   \n",
       "2             A2M  ENSG00000175899  protein_coding    9115927      -   \n",
       "3           A2ML1  ENSG00000166535  protein_coding    8822627      +   \n",
       "4          A4GALT  ENSG00000128274  protein_coding   42720879      -   \n",
       "...           ...              ...             ...        ...    ...   \n",
       "15615      ZYG11A  ENSG00000203995  protein_coding   52842770      +   \n",
       "15616      ZYG11B  ENSG00000162378  protein_coding   52726478      +   \n",
       "15617         ZYX  ENSG00000159840  protein_coding  143381361      +   \n",
       "15618       ZZEF1  ENSG00000074755  protein_coding    4143055      -   \n",
       "15619        ZZZ3  ENSG00000036549  protein_coding   77683457      -   \n",
       "\n",
       "      chromosome_name_y  \n",
       "0                 chr19  \n",
       "1                 chr10  \n",
       "2                 chr12  \n",
       "3                 chr12  \n",
       "4                 chr22  \n",
       "...                 ...  \n",
       "15615              chr1  \n",
       "15616              chr1  \n",
       "15617              chr7  \n",
       "15618             chr17  \n",
       "15619              chr1  \n",
       "\n",
       "[15620 rows x 15 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "newDataset1_0 = pd.read_hdf(\"../CTB/CTB_128k_slack_0.h5\", key=\"df\", mode=\"r\")\n",
    "newDataset1_1 = pd.read_hdf(\"../CTB/CTB_128k_slack_1.h5\", key=\"df\", mode=\"r\")\n",
    "newDataset1_2 = pd.read_hdf(\"../CTB/CTB_128k_slack_2.h5\", key=\"df\", mode=\"r\")\n",
    "newDataset1_3 = pd.read_hdf(\"../CTB/CTB_128k_slack_3.h5\", key=\"df\", mode=\"r\")\n",
    "newDataset1_4 = pd.read_hdf(\"../CTB/CTB_128k_slack_4.h5\", key=\"df\", mode=\"r\")\n",
    "newDataset1_5 = pd.read_hdf(\"../CTB/CTB_128k_slack_5.h5\", key=\"df\", mode=\"r\")\n",
    "newDataset1_6 = pd.read_hdf(\"../CTB/CTB_128k_slack_6.h5\", key=\"df\", mode=\"r\")\n",
    "\n",
    "dataset = pd.concat([\n",
    "                            newDataset1_0,\n",
    "                            newDataset1_1,\n",
    "                            newDataset1_2,\n",
    "                            newDataset1_3,\n",
    "                            newDataset1_4,\n",
    "                            newDataset1_5,\n",
    "                            newDataset1_6,\n",
    "                         ])\n",
    "\n",
    "\n",
    "#join dei due dataframe su gene_name\n",
    "df2 = pd.merge(dataset, lizio, on=\"gene_id\", how=\"inner\")\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prendi tutti i gene_id in dataset che non sono in lizio\n",
    "geni_mancanti = dataset[~dataset.gene_id.isin(lizio.gene_id)]\n",
    "geni_mancanti =geni_mancanti['gene_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Danilo\\AppData\\Local\\Temp\\ipykernel_10648\\3157581289.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2.drop(columns=[\"strand_y\"],inplace=True)\n",
      "C:\\Users\\Danilo\\AppData\\Local\\Temp\\ipykernel_10648\\3157581289.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2.rename(columns={\"strand_x\":\"strand\"},inplace=True)\n",
      "C:\\Users\\Danilo\\AppData\\Local\\Temp\\ipykernel_10648\\3157581289.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2.dropna(inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_stable_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>TSS</th>\n",
       "      <th>chromosome_name_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58858938</td>\n",
       "      <td>58859039</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>ENSG00000121410.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>58347673</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>52645379</td>\n",
       "      <td>52645393</td>\n",
       "      <td>-</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>ENSG00000148584.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50885633</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>9268507</td>\n",
       "      <td>9268523</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>ENSG00000175899.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>9115927</td>\n",
       "      <td>chr12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>8975206</td>\n",
       "      <td>8975223</td>\n",
       "      <td>+</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "      <td>ENSG00000166535.20</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>8822627</td>\n",
       "      <td>chr12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22</td>\n",
       "      <td>43116867</td>\n",
       "      <td>43116885</td>\n",
       "      <td>-</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>ENSG00000128274.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>42720879</td>\n",
       "      <td>chr22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15615</th>\n",
       "      <td>chr1</td>\n",
       "      <td>53308425</td>\n",
       "      <td>53308442</td>\n",
       "      <td>+</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>ENSG00000203995.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>52842770</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15616</th>\n",
       "      <td>chr1</td>\n",
       "      <td>53192114</td>\n",
       "      <td>53192150</td>\n",
       "      <td>+</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>ENSG00000162378.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>52726478</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15617</th>\n",
       "      <td>chr7</td>\n",
       "      <td>143078379</td>\n",
       "      <td>143078454</td>\n",
       "      <td>+</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>ENSG00000159840.16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>143381361</td>\n",
       "      <td>chr7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15618</th>\n",
       "      <td>chr17</td>\n",
       "      <td>4046257</td>\n",
       "      <td>4046349</td>\n",
       "      <td>-</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>ENSG00000074755.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4143055</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15619</th>\n",
       "      <td>chr1</td>\n",
       "      <td>78149041</td>\n",
       "      <td>78149142</td>\n",
       "      <td>-</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>77683457</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15584 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome_name_x      start        end strand gene_name  \\\n",
       "0                 chr19   58858938   58859039      -      A1BG   \n",
       "1                 chr10   52645379   52645393      -      A1CF   \n",
       "2                 chr12    9268507    9268523      -       A2M   \n",
       "3                 chr12    8975206    8975223      +     A2ML1   \n",
       "4                 chr22   43116867   43116885      -    A4GALT   \n",
       "...                 ...        ...        ...    ...       ...   \n",
       "15615              chr1   53308425   53308442      +    ZYG11A   \n",
       "15616              chr1   53192114   53192150      +    ZYG11B   \n",
       "15617              chr7  143078379  143078454      +       ZYX   \n",
       "15618             chr17    4046257    4046349      -     ZZEF1   \n",
       "15619              chr1   78149041   78149142      -      ZZZ3   \n",
       "\n",
       "        gene_stable_id             gene_id       gene_type        TSS  \\\n",
       "0      ENSG00000121410  ENSG00000121410.12  protein_coding   58347673   \n",
       "1      ENSG00000148584  ENSG00000148584.15  protein_coding   50885633   \n",
       "2      ENSG00000175899  ENSG00000175899.15  protein_coding    9115927   \n",
       "3      ENSG00000166535  ENSG00000166535.20  protein_coding    8822627   \n",
       "4      ENSG00000128274  ENSG00000128274.17  protein_coding   42720879   \n",
       "...                ...                 ...             ...        ...   \n",
       "15615  ENSG00000203995  ENSG00000203995.10  protein_coding   52842770   \n",
       "15616  ENSG00000162378  ENSG00000162378.13  protein_coding   52726478   \n",
       "15617  ENSG00000159840  ENSG00000159840.16  protein_coding  143381361   \n",
       "15618  ENSG00000074755  ENSG00000074755.15  protein_coding    4143055   \n",
       "15619  ENSG00000036549  ENSG00000036549.13  protein_coding   77683457   \n",
       "\n",
       "      chromosome_name_y  \n",
       "0                 chr19  \n",
       "1                 chr10  \n",
       "2                 chr12  \n",
       "3                 chr12  \n",
       "4                 chr22  \n",
       "...                 ...  \n",
       "15615              chr1  \n",
       "15616              chr1  \n",
       "15617              chr7  \n",
       "15618             chr17  \n",
       "15619              chr1  \n",
       "\n",
       "[15584 rows x 10 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2=df2[df2['strand_x']==df2['strand_y']]\n",
    "df2.drop(columns=[\"strand_y\"],inplace=True)\n",
    "#rinominazione colonne\n",
    "df2.rename(columns={\"strand_x\":\"strand\"},inplace=True)\n",
    "#drop nan\n",
    "df2.dropna(inplace=True)\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Transcription start site (TSS)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Gene type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000261657.1</td>\n",
       "      <td>66119285</td>\n",
       "      <td>HG991_PATCH</td>\n",
       "      <td>+</td>\n",
       "      <td>SLC25A26</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000261657.1</td>\n",
       "      <td>66320895</td>\n",
       "      <td>HG991_PATCH</td>\n",
       "      <td>+</td>\n",
       "      <td>SLC25A26</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000261657.1</td>\n",
       "      <td>66320895</td>\n",
       "      <td>HG991_PATCH</td>\n",
       "      <td>+</td>\n",
       "      <td>SLC25A26</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000261657.1</td>\n",
       "      <td>66320895</td>\n",
       "      <td>HG991_PATCH</td>\n",
       "      <td>+</td>\n",
       "      <td>SLC25A26</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000261657.1</td>\n",
       "      <td>66339287</td>\n",
       "      <td>HG991_PATCH</td>\n",
       "      <td>+</td>\n",
       "      <td>SLC25A26</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215399</th>\n",
       "      <td>ENSG00000263267.1</td>\n",
       "      <td>198858</td>\n",
       "      <td>HSCHR17_1_CTG1</td>\n",
       "      <td>+</td>\n",
       "      <td>RP11-1260E13.4</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215400</th>\n",
       "      <td>ENSG00000263267.1</td>\n",
       "      <td>199139</td>\n",
       "      <td>HSCHR17_1_CTG1</td>\n",
       "      <td>+</td>\n",
       "      <td>RP11-1260E13.4</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215401</th>\n",
       "      <td>ENSG00000263267.1</td>\n",
       "      <td>199766</td>\n",
       "      <td>HSCHR17_1_CTG1</td>\n",
       "      <td>+</td>\n",
       "      <td>RP11-1260E13.4</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215402</th>\n",
       "      <td>ENSG00000262336.1</td>\n",
       "      <td>220262</td>\n",
       "      <td>HSCHR17_1_CTG1</td>\n",
       "      <td>+</td>\n",
       "      <td>RP11-1260E13.2</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215403</th>\n",
       "      <td>ENSG00000262005.1</td>\n",
       "      <td>222361</td>\n",
       "      <td>HSCHR17_1_CTG1</td>\n",
       "      <td>+</td>\n",
       "      <td>RP11-1260E13.3</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>215404 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id  Transcription start site (TSS)  \\\n",
       "0       ENSG00000261657.1                        66119285   \n",
       "1       ENSG00000261657.1                        66320895   \n",
       "2       ENSG00000261657.1                        66320895   \n",
       "3       ENSG00000261657.1                        66320895   \n",
       "4       ENSG00000261657.1                        66339287   \n",
       "...                   ...                             ...   \n",
       "215399  ENSG00000263267.1                          198858   \n",
       "215400  ENSG00000263267.1                          199139   \n",
       "215401  ENSG00000263267.1                          199766   \n",
       "215402  ENSG00000262336.1                          220262   \n",
       "215403  ENSG00000262005.1                          222361   \n",
       "\n",
       "       Chromosome/scaffold name strand       gene_name             Gene type  \n",
       "0                   HG991_PATCH      +        SLC25A26        protein_coding  \n",
       "1                   HG991_PATCH      +        SLC25A26        protein_coding  \n",
       "2                   HG991_PATCH      +        SLC25A26        protein_coding  \n",
       "3                   HG991_PATCH      +        SLC25A26        protein_coding  \n",
       "4                   HG991_PATCH      +        SLC25A26        protein_coding  \n",
       "...                         ...    ...             ...                   ...  \n",
       "215399           HSCHR17_1_CTG1      +  RP11-1260E13.4  processed_transcript  \n",
       "215400           HSCHR17_1_CTG1      +  RP11-1260E13.4  processed_transcript  \n",
       "215401           HSCHR17_1_CTG1      +  RP11-1260E13.4  processed_transcript  \n",
       "215402           HSCHR17_1_CTG1      +  RP11-1260E13.2             antisense  \n",
       "215403           HSCHR17_1_CTG1      +  RP11-1260E13.3             antisense  \n",
       "\n",
       "[215404 rows x 6 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meging=pd.read_csv(\"..\\\\dataset\\\\ens37_mancanti.txt\", sep=\"\\t\")\n",
    "\n",
    "meging.dropna(inplace=True)\n",
    "#modifica la colonna Strand assegnando - se e uguale a -1 e + se e uguale a 1\n",
    "meging['strand'] = meging['strand'].apply(lambda x: \"-\" if x == -1 else \"+\")\n",
    "meging\n",
    "meging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "ENSG00000205336.7     82\n",
       "ENSG00000165795.16    78\n",
       "ENSG00000161249.16    69\n",
       "ENSG00000154556.13    65\n",
       "ENSG00000179818.9     63\n",
       "                      ..\n",
       "ENSG00000254352.1      1\n",
       "ENSG00000216073.2      1\n",
       "ENSG00000273245.1      1\n",
       "ENSG00000268681.1      1\n",
       "ENSG00000262005.1      1\n",
       "Name: count, Length: 63677, dtype: int64"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meging['gene_id'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Transcription start site (TSS)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Gene type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8299</th>\n",
       "      <td>ENSG00000232263.1</td>\n",
       "      <td>31661832</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>KRTAP25-1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8350</th>\n",
       "      <td>ENSG00000206107.2</td>\n",
       "      <td>31710012</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>KRTAP27-1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8423</th>\n",
       "      <td>ENSG00000240432.3</td>\n",
       "      <td>31798286</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>KRTAP13-3</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8594</th>\n",
       "      <td>ENSG00000186925.6</td>\n",
       "      <td>31914183</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>KRTAP19-6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8639</th>\n",
       "      <td>ENSG00000206106.2</td>\n",
       "      <td>31962716</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>KRTAP22-2</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199912</th>\n",
       "      <td>ENSG00000180016.2</td>\n",
       "      <td>3301704</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>OR1E1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199923</th>\n",
       "      <td>ENSG00000127780.3</td>\n",
       "      <td>3337135</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>OR1E2</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199974</th>\n",
       "      <td>ENSG00000125650.4</td>\n",
       "      <td>6379069</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>PSPN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199975</th>\n",
       "      <td>ENSG00000125650.4</td>\n",
       "      <td>6375860</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>PSPN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201313</th>\n",
       "      <td>ENSG00000186306.1</td>\n",
       "      <td>158369256</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>OR10T2</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>256 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id  Transcription start site (TSS)  \\\n",
       "8299    ENSG00000232263.1                        31661832   \n",
       "8350    ENSG00000206107.2                        31710012   \n",
       "8423    ENSG00000240432.3                        31798286   \n",
       "8594    ENSG00000186925.6                        31914183   \n",
       "8639    ENSG00000206106.2                        31962716   \n",
       "...                   ...                             ...   \n",
       "199912  ENSG00000180016.2                         3301704   \n",
       "199923  ENSG00000127780.3                         3337135   \n",
       "199974  ENSG00000125650.4                         6379069   \n",
       "199975  ENSG00000125650.4                         6375860   \n",
       "201313  ENSG00000186306.1                       158369256   \n",
       "\n",
       "       Chromosome/scaffold name strand  gene_name       Gene type  \n",
       "8299                         21      -  KRTAP25-1  protein_coding  \n",
       "8350                         21      -  KRTAP27-1  protein_coding  \n",
       "8423                         21      -  KRTAP13-3  protein_coding  \n",
       "8594                         21      -  KRTAP19-6  protein_coding  \n",
       "8639                         21      -  KRTAP22-2  protein_coding  \n",
       "...                         ...    ...        ...             ...  \n",
       "199912                       17      -      OR1E1  protein_coding  \n",
       "199923                       17      -      OR1E2  protein_coding  \n",
       "199974                       19      -       PSPN  protein_coding  \n",
       "199975                       19      -       PSPN  protein_coding  \n",
       "201313                        1      -     OR10T2  protein_coding  \n",
       "\n",
       "[256 rows x 6 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#seleziona i gene_id in meging che sono in df3\n",
    "df4 = meging[meging.gene_id.isin(geni_mancanti)]\n",
    "df4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>color</th>\n",
       "      <th>strand</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>RGB</th>\n",
       "      <th>p</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>fpkm_median</th>\n",
       "      <th>tpm_median</th>\n",
       "      <th>fpkm_uq_median</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>p2@LOXL4,0.1291</td>\n",
       "      <td>55.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p2</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "      <td>1.800827</td>\n",
       "      <td>3.202559</td>\n",
       "      <td>1.977610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>p3@LOXL4,0.1842</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>30,144,255</td>\n",
       "      <td>p3</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "      <td>1.800827</td>\n",
       "      <td>3.202559</td>\n",
       "      <td>1.977610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>p1@LOXL4,0.2200</td>\n",
       "      <td>48.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>30,144,255</td>\n",
       "      <td>p1</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "      <td>1.800827</td>\n",
       "      <td>3.202559</td>\n",
       "      <td>1.977610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>p1@PYROXD2,0.2721</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>PYROXD2</td>\n",
       "      <td>ENSG00000119943.13</td>\n",
       "      <td>1.167486</td>\n",
       "      <td>2.608927</td>\n",
       "      <td>1.382999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174957</td>\n",
       "      <td>100174982</td>\n",
       "      <td>p2@PYROXD2,0.2448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100174957</td>\n",
       "      <td>100174982</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p2</td>\n",
       "      <td>PYROXD2</td>\n",
       "      <td>ENSG00000119943.13</td>\n",
       "      <td>1.167486</td>\n",
       "      <td>2.608927</td>\n",
       "      <td>1.382999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153925</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2870826</td>\n",
       "      <td>2870844</td>\n",
       "      <td>p2@ZFY-AS1,0.0950</td>\n",
       "      <td>-159.0</td>\n",
       "      <td>-</td>\n",
       "      <td>2870826</td>\n",
       "      <td>2870844</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p2</td>\n",
       "      <td>ZFY-AS1</td>\n",
       "      <td>ENSG00000233070.1</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153926</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778505</td>\n",
       "      <td>6778534</td>\n",
       "      <td>p6@TBL1Y,0.3422</td>\n",
       "      <td>-192.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778505</td>\n",
       "      <td>6778534</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p6</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153927</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778592</td>\n",
       "      <td>6778601</td>\n",
       "      <td>p7@TBL1Y,0.4173</td>\n",
       "      <td>-125.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778592</td>\n",
       "      <td>6778601</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p7</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153928</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778602</td>\n",
       "      <td>6778616</td>\n",
       "      <td>p4@TBL1Y,0.4601</td>\n",
       "      <td>-110.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778602</td>\n",
       "      <td>6778616</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p4</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153929</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778870</td>\n",
       "      <td>6778873</td>\n",
       "      <td>p9@TBL1Y,0.4948</td>\n",
       "      <td>132.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778870</td>\n",
       "      <td>6778873</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p9</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>153930 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome_name      start        end           locus_id  color strand  \\\n",
       "0                chr10  100015362  100015397    p2@LOXL4,0.1291   55.0      -   \n",
       "1                chr10  100017518  100017519    p3@LOXL4,0.1842   13.0      -   \n",
       "2                chr10  100027943  100027958    p1@LOXL4,0.2200   48.0      -   \n",
       "3                chr10  100174900  100174956  p1@PYROXD2,0.2721    0.0      -   \n",
       "4                chr10  100174957  100174982  p2@PYROXD2,0.2448    0.0      -   \n",
       "...                ...        ...        ...                ...    ...    ...   \n",
       "153925            chrY    2870826    2870844  p2@ZFY-AS1,0.0950 -159.0      -   \n",
       "153926            chrY    6778505    6778534    p6@TBL1Y,0.3422 -192.0      +   \n",
       "153927            chrY    6778592    6778601    p7@TBL1Y,0.4173 -125.0      +   \n",
       "153928            chrY    6778602    6778616    p4@TBL1Y,0.4601 -110.0      +   \n",
       "153929            chrY    6778870    6778873    p9@TBL1Y,0.4948  132.0      +   \n",
       "\n",
       "           start2       end2          RGB   p gene_name             gene_id  \\\n",
       "0       100015362  100015397  211,211,211  p2     LOXL4   ENSG00000138131.4   \n",
       "1       100017518  100017519   30,144,255  p3     LOXL4   ENSG00000138131.4   \n",
       "2       100027943  100027958   30,144,255  p1     LOXL4   ENSG00000138131.4   \n",
       "3       100174900  100174956   60,179,113  p1   PYROXD2  ENSG00000119943.13   \n",
       "4       100174957  100174982   60,179,113  p2   PYROXD2  ENSG00000119943.13   \n",
       "...           ...        ...          ...  ..       ...                 ...   \n",
       "153925    2870826    2870844  211,211,211  p2   ZFY-AS1   ENSG00000233070.1   \n",
       "153926    6778505    6778534   60,179,113  p6     TBL1Y  ENSG00000092377.15   \n",
       "153927    6778592    6778601   60,179,113  p7     TBL1Y  ENSG00000092377.15   \n",
       "153928    6778602    6778616   60,179,113  p4     TBL1Y  ENSG00000092377.15   \n",
       "153929    6778870    6778873   60,179,113  p9     TBL1Y  ENSG00000092377.15   \n",
       "\n",
       "        fpkm_median  tpm_median  fpkm_uq_median  \n",
       "0          1.800827    3.202559        1.977610  \n",
       "1          1.800827    3.202559        1.977610  \n",
       "2          1.800827    3.202559        1.977610  \n",
       "3          1.167486    2.608927        1.382999  \n",
       "4          1.167486    2.608927        1.382999  \n",
       "...             ...         ...             ...  \n",
       "153925    -7.643856   -7.643856       -7.643856  \n",
       "153926    -7.643856   -7.643856       -7.643856  \n",
       "153927    -7.643856   -7.643856       -7.643856  \n",
       "153928    -7.643856   -7.643856       -7.643856  \n",
       "153929    -7.643856   -7.643856       -7.643856  \n",
       "\n",
       "[153930 rows x 15 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient=pd.read_csv(\"..\\\\dataset\\\\Dataset_median.csv\", sep=\",\")\n",
    "#merge su gene_id\n",
    "df2 = pd.merge(df1, patient, on=\"gene_id\")\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_name\n",
       "MEG3         176\n",
       "SLC1A2       170\n",
       "PAPPA        153\n",
       "APOB         147\n",
       "QKI          137\n",
       "            ... \n",
       "PAPPA-AS1      1\n",
       "MRPL50P2       1\n",
       "RPS26P49       1\n",
       "KRT8P15        1\n",
       "ANKRD36P1      1\n",
       "Name: count, Length: 14198, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[\"gene_name\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "p\n",
       "p1    14075\n",
       "p2       86\n",
       "p3       14\n",
       "p4       10\n",
       "p5        5\n",
       "p6        3\n",
       "p7        2\n",
       "p8        2\n",
       "p9        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df3 = df2.loc[df2.groupby(\"gene_name\")[\"p_value\"].idxmin()]\n",
    "\n",
    "\n",
    "\n",
    "df3.value_counts(\"p\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>color</th>\n",
       "      <th>strand</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>RGB</th>\n",
       "      <th>p</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>fpkm_median</th>\n",
       "      <th>tpm_median</th>\n",
       "      <th>fpkm_uq_median</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>153866</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99839925</td>\n",
       "      <td>99839938</td>\n",
       "      <td>p1@TNMD,0.1353</td>\n",
       "      <td>126.0</td>\n",
       "      <td>+</td>\n",
       "      <td>99839925</td>\n",
       "      <td>99839938</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p1</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>-3.925769</td>\n",
       "      <td>-2.475535</td>\n",
       "      <td>-3.733124</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38903</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169863016</td>\n",
       "      <td>169863109</td>\n",
       "      <td>p1@SCYL3,0.2639</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>169863016</td>\n",
       "      <td>169863109</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>0.873892</td>\n",
       "      <td>2.396543</td>\n",
       "      <td>1.063089</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42198</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27961629</td>\n",
       "      <td>27961665</td>\n",
       "      <td>p1@FGR,0.5037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>27961629</td>\n",
       "      <td>27961665</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>FGR</td>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "      <td>0.612022</td>\n",
       "      <td>2.081715</td>\n",
       "      <td>0.757621</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71796</th>\n",
       "      <td>chr6</td>\n",
       "      <td>143832820</td>\n",
       "      <td>143832875</td>\n",
       "      <td>p1@FUCA2,0.4000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>143832820</td>\n",
       "      <td>143832875</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>FUCA2</td>\n",
       "      <td>ENSG00000001036.14</td>\n",
       "      <td>3.827545</td>\n",
       "      <td>5.333456</td>\n",
       "      <td>4.001946</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74636</th>\n",
       "      <td>chr6</td>\n",
       "      <td>53409890</td>\n",
       "      <td>53409906</td>\n",
       "      <td>p1@GCLC,0.4228</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-</td>\n",
       "      <td>53409890</td>\n",
       "      <td>53409906</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>GCLC</td>\n",
       "      <td>ENSG00000001084.13</td>\n",
       "      <td>1.509544</td>\n",
       "      <td>2.956224</td>\n",
       "      <td>1.694702</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86592</th>\n",
       "      <td>chrX</td>\n",
       "      <td>101967366</td>\n",
       "      <td>101967391</td>\n",
       "      <td>p1@ARMCX5-GPRASP2,p1@GPRASP2,0.6351</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>101967366</td>\n",
       "      <td>101967391</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>ARMCX5-GPRASP2</td>\n",
       "      <td>ENSG00000286237.1</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65961</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89513578</td>\n",
       "      <td>89513670</td>\n",
       "      <td>p1@HERC3,0.3732</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>89513578</td>\n",
       "      <td>89513670</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>HERC3</td>\n",
       "      <td>ENSG00000287542.1</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82689</th>\n",
       "      <td>chr8</td>\n",
       "      <td>67579807</td>\n",
       "      <td>67579860</td>\n",
       "      <td>p1@C8orf44,p1@SGK3,0.4647</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>67579807</td>\n",
       "      <td>67579860</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>C8orf44</td>\n",
       "      <td>ENSG00000288596.2</td>\n",
       "      <td>0.995448</td>\n",
       "      <td>2.458933</td>\n",
       "      <td>1.200316</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148900</th>\n",
       "      <td>chr8</td>\n",
       "      <td>53852965</td>\n",
       "      <td>53852990</td>\n",
       "      <td>p1@NPBWR1,0.2996</td>\n",
       "      <td>497.0</td>\n",
       "      <td>+</td>\n",
       "      <td>53852965</td>\n",
       "      <td>53852990</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>NPBWR1</td>\n",
       "      <td>ENSG00000288611.1</td>\n",
       "      <td>-6.687800</td>\n",
       "      <td>-5.779918</td>\n",
       "      <td>-6.546245</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87371</th>\n",
       "      <td>chrX</td>\n",
       "      <td>139866699</td>\n",
       "      <td>139866741</td>\n",
       "      <td>p1@CDR1,0.6297</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>139866699</td>\n",
       "      <td>139866741</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>CDR1</td>\n",
       "      <td>ENSG00000288642.1</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14272 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome_name      start        end  \\\n",
       "153866            chrX   99839925   99839938   \n",
       "38903             chr1  169863016  169863109   \n",
       "42198             chr1   27961629   27961665   \n",
       "71796             chr6  143832820  143832875   \n",
       "74636             chr6   53409890   53409906   \n",
       "...                ...        ...        ...   \n",
       "86592             chrX  101967366  101967391   \n",
       "65961             chr4   89513578   89513670   \n",
       "82689             chr8   67579807   67579860   \n",
       "148900            chr8   53852965   53852990   \n",
       "87371             chrX  139866699  139866741   \n",
       "\n",
       "                                   locus_id  color strand     start2  \\\n",
       "153866                       p1@TNMD,0.1353  126.0      +   99839925   \n",
       "38903                       p1@SCYL3,0.2639    0.0      -  169863016   \n",
       "42198                         p1@FGR,0.5037    0.0      -   27961629   \n",
       "71796                       p1@FUCA2,0.4000    0.0      -  143832820   \n",
       "74636                        p1@GCLC,0.4228   -4.0      -   53409890   \n",
       "...                                     ...    ...    ...        ...   \n",
       "86592   p1@ARMCX5-GPRASP2,p1@GPRASP2,0.6351    0.0      +  101967366   \n",
       "65961                       p1@HERC3,0.3732    0.0      +   89513578   \n",
       "82689             p1@C8orf44,p1@SGK3,0.4647    0.0      +   67579807   \n",
       "148900                     p1@NPBWR1,0.2996  497.0      +   53852965   \n",
       "87371                        p1@CDR1,0.6297    0.0      -  139866699   \n",
       "\n",
       "             end2          RGB   p       gene_name             gene_id  \\\n",
       "153866   99839938  211,211,211  p1            TNMD   ENSG00000000005.6   \n",
       "38903   169863109   60,179,113  p1           SCYL3  ENSG00000000457.14   \n",
       "42198    27961665   60,179,113  p1             FGR  ENSG00000000938.13   \n",
       "71796   143832875   60,179,113  p1           FUCA2  ENSG00000001036.14   \n",
       "74636    53409906   60,179,113  p1            GCLC  ENSG00000001084.13   \n",
       "...           ...          ...  ..             ...                 ...   \n",
       "86592   101967391   60,179,113  p1  ARMCX5-GPRASP2   ENSG00000286237.1   \n",
       "65961    89513670   60,179,113  p1           HERC3   ENSG00000287542.1   \n",
       "82689    67579860   60,179,113  p1         C8orf44   ENSG00000288596.2   \n",
       "148900   53852990   60,179,113  p1          NPBWR1   ENSG00000288611.1   \n",
       "87371   139866741   60,179,113  p1            CDR1   ENSG00000288642.1   \n",
       "\n",
       "        fpkm_median  tpm_median  fpkm_uq_median  p_value  \n",
       "153866    -3.925769   -2.475535       -3.733124        1  \n",
       "38903      0.873892    2.396543        1.063089        1  \n",
       "42198      0.612022    2.081715        0.757621        1  \n",
       "71796      3.827545    5.333456        4.001946        1  \n",
       "74636      1.509544    2.956224        1.694702        1  \n",
       "...             ...         ...             ...      ...  \n",
       "86592     -7.643856   -7.643856       -7.643856        1  \n",
       "65961     -7.643856   -7.643856       -7.643856        1  \n",
       "82689      0.995448    2.458933        1.200316        1  \n",
       "148900    -6.687800   -5.779918       -6.546245        1  \n",
       "87371     -7.643856   -7.643856       -7.643856        1  \n",
       "\n",
       "[14272 rows x 16 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df4 = df2.loc[df2.groupby(\"gene_id\")[\"p_value\"].idxmin()]\n",
    "df4"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "AIE",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
