{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:8: SyntaxWarning: invalid escape sequence '\\G'\n",
      "<>:8: SyntaxWarning: invalid escape sequence '\\G'\n",
      "C:\\Users\\dany0\\AppData\\Local\\Temp\\ipykernel_14800\\4216738839.py:8: SyntaxWarning: invalid escape sequence '\\G'\n",
      "  for i, record in enumerate(SeqIO.parse(\"..\\\\dataset\\\\\\GRCh37.primary_assembly.genome.fa\", \"fasta\")):\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "chrdict = {}\n",
    "\n",
    "for i, record in enumerate(SeqIO.parse(\"..\\\\dataset\\\\\\GRCh37.primary_assembly.genome.fa\", \"fasta\")):\n",
    "    chrdict[record.id] = record.seq\n",
    "\n",
    "\n",
    "# LE SE SEQUENZE DI BASI AZOTATE SONO MEMORIZZATE IN UN DIZIONARIO CHE USA COME CHIAVE \"chr#\" dove # indica\n",
    "# il numero del cromosoma oppure X/Y/M (M non ci interessa perché é il DNA mitocondriale che non viene usato per questo caso)\n",
    "\n",
    "\n",
    "# UA INIZIA INVECE LA PARTE DI LETTURA DEL FILE CONTENENTE I TSS\n",
    "\n",
    "\n",
    "def extract_p(p):\n",
    "    return p.split(\"@\")[0]\n",
    "\n",
    "def extract_gene_name(string):\n",
    "    string =  string.split(\"@\")[1]\n",
    "    return string.split(\",\")[0]\n",
    "\n",
    "v1 = np.vectorize(extract_p)\n",
    "v2 = np.vectorize(extract_gene_name)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr10': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr11': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr12': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr13': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr14': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr15': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr16': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr17': Seq('AAGCTTCTCACCCTGTTCCTGCATAGATAATTGCATGACAATTGCCTTGTCCCT...GGT'),\n",
       " 'chr18': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr19': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr2': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr20': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr21': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr22': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr3': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr4': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr5': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr6': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr7': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr8': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chr9': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chrM': Seq('GATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGG...ATG'),\n",
       " 'chrX': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'chrY': Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN'),\n",
       " 'GL000192.1': Seq('GAATTCATTCACCATTATTCTTTTATAATATTGCTATTTTATTATTCTTGATCA...TTC'),\n",
       " 'GL000225.1': Seq('GAATTCTCACATGTCTCTGCACTGATCACCCAGGTGATGTAACTCCTGTCTAGG...TTC'),\n",
       " 'GL000194.1': Seq('GATCCCTGCCCTAAAACTTTCCCCCCTCATGTCCAGCAAATGCTGCATGGAGCC...ATC'),\n",
       " 'GL000193.1': Seq('GAATTCATTTCTTTGATTTGTAGAGTGGCAGGGTACCTAAAGTTAACATTTGTC...TTC'),\n",
       " 'GL000200.1': Seq('TGGGTTGATAGGTTTTTAAGAATTCTTTTATGACTTCAAAAGTGTCACTTCGTT...CCT'),\n",
       " 'GL000222.1': Seq('GAATTCTATGTGAGGGACACACATTCAAAACCCAGCAGCAGTATTCTGGAATCC...TTC'),\n",
       " 'GL000212.1': Seq('GAATTCCTCCTTGCACATTTCTGCTTTATCTCATCTACACAGAAGTGATCCAAT...TTC'),\n",
       " 'GL000195.1': Seq('GAATTCCTCGTTCACACAGTTTCTTAAGCTTCCTGGGATGCGACCTGTGATGGC...TTC'),\n",
       " 'GL000223.1': Seq('GAATTCTCTGATGTCGAGTGAGGTTGGAGGCACGGCTAAAGGCCTTTCCACACT...TTC'),\n",
       " 'GL000224.1': Seq('GAATTCTGTTGATGATGGATGTGAGAGTGGTGAAGCCCTGCGGGATGATGTAGA...TTC'),\n",
       " 'GL000219.1': Seq('GAATTCTACCTGTGGGACAAACATTGAGACCTTTGTAACATTGTTCTGGAATGC...TTC'),\n",
       " 'GL000205.1': Seq('GAATTCTTGGTATTATTCCTTTTGAGTACCCCTATTTATGAGAGAGAAAACGAA...TTC'),\n",
       " 'GL000215.1': Seq('GAATTCCATAATATAAAGTTACTTTTTTAAAAAAAAATTAGACAATCAATTATC...TTC'),\n",
       " 'GL000216.1': Seq('GAATTCCATTGCATTCCATTCCATTTCATTCTATTACACTCCATTCCATTCCTT...TTC'),\n",
       " 'GL000217.1': Seq('GAATTCACTGGTCTTACCATATTCCCCATCTTCCTGAAGCAGCTGGATTGATAG...TTC'),\n",
       " 'GL000199.1': Seq('GAATTCTATATGAGGGAGAAAAACTCAGAACCTAGCAGCAGTGTTCTGGAATCC...TTC'),\n",
       " 'GL000211.1': Seq('GAATTCTTTTTCAAGTAAGTCAGGGATTTCTTCTTAGTTTGGATCCATTGCTGG...TTC'),\n",
       " 'GL000213.1': Seq('GAATTCAGTGAATGAGCCAAGATTCTCCAGAAACCGCTCAGCTTTGCAGTGATA...TTC'),\n",
       " 'GL000220.1': Seq('GAATTCAATACTTAAACATAAAACCTAATAAAATTCTTAGAAGTAAACATAAGG...TTC'),\n",
       " 'GL000218.1': Seq('GAATTCATGTAAAGAAAGACCATGGGTGTTTGTACTGGATTTAGTATTCATCAT...TTC'),\n",
       " 'GL000209.1': Seq('GATCAGCTTTCCCTTTGAAATCAGAGTACTAATCTGACTCATTGTTCCCTGAAT...ATC'),\n",
       " 'GL000221.1': Seq('GAATTCGTTGAGGAGCTTCTGGAAAGTGCACATTCTGACTCAGCAGGTATTGGA...TTC'),\n",
       " 'GL000214.1': Seq('GAATTCAGTGCAATTCAAAACAGTAATTTGTGGGAAAAATTAGATATGCAGATA...TTC'),\n",
       " 'GL000228.1': Seq('GAATTCTGTGCATTTTGTTTGATGGAATTTGGTATGTTTACCAAAATGGCAGAA...TTC'),\n",
       " 'GL000227.1': Seq('GATCCAACTATATTGTCCACAAGATACTCACTTTAGATCAAAATACACAATGAG...ATC'),\n",
       " 'GL000191.1': Seq('GATCCACCTGCCTCAGCCTCCCAGAGTGCTGGGATTATAGGTGTGAGCCACTGC...ATC'),\n",
       " 'GL000208.1': Seq('GAATTCTTCAAAGAGTTCCAGATATCCACAGGCAGATTCTACAAAAGAAGTGTT...GAT'),\n",
       " 'GL000198.1': Seq('GAATTCGTCTCATAGAGTTGAACCTTTCTTTTCATTGATCAGTTTGGAAACACT...TTC'),\n",
       " 'GL000204.1': Seq('GATCAGGAAGAGTCCTGGGGGGCGGGGGGCGAGCACAGGCAGTGGAAGGGAAGC...ATC'),\n",
       " 'GL000233.1': Seq('GATCCCATTTGTCAATTTTGGATTTTGTGGCCATTGCTTCTGTTGTTTTAGACA...ATC'),\n",
       " 'GL000237.1': Seq('GATCTCAGGAGGCGGAGGTTGCAGTGATCGGTGGCCATAGAAGCCCTCCTGAGC...ATC'),\n",
       " 'GL000230.1': Seq('GATCCTTTGAGCCCAGGAGTTTGAGACCACCCTGGGAAACATGGCAAAACCCTC...ATC'),\n",
       " 'GL000242.1': Seq('GATCATGCCTGGGACACACAGCAAGACCTTGTCCCTTAAAAAAAAAAAAAAAAA...ATC'),\n",
       " 'GL000243.1': Seq('GATCTTCAGGCTCTGTCTCTGATTCTAGTGCTCTTGTTATTTCCACCATATCTG...ATC'),\n",
       " 'GL000241.1': Seq('GATCTGAAGACAAATCAGGGTTGCAGGATGGTTAAATATCTGCAAGTCAATAAA...ATC'),\n",
       " 'GL000236.1': Seq('GATCACATGGACACAGGAAGGGGAATATCACACTCTGGGGACTGTTGTGGGGTG...ATC'),\n",
       " 'GL000240.1': Seq('GATCAGTGGAGACCTGGTCAGCTGGGGCTTAGTGCTGGCCTGGTCAGCATGGGC...ATC'),\n",
       " 'GL000206.1': Seq('GATCCCACCAGAGGCCATAAACAGATAGTTCAGAGCTATGAGAACCCAGGTGGC...ATC'),\n",
       " 'GL000232.1': Seq('GATCAGGCCCAGGAACTGCTGCCTTGGGTAGCTTTGGAACAAGGCTCCTGCATC...ATC'),\n",
       " 'GL000234.1': Seq('GATCTCAATGCATCCAGATTCATAATATGTTTGCCTTTGAATGGAATGATCCAG...ATC'),\n",
       " 'GL000202.1': Seq('TGCCAGAAAAGGGTGTGTGTTGGAGGGGGTGTCTTTCACAATATATTTTTAATT...AGC'),\n",
       " 'GL000238.1': Seq('GATCACAGTATCACAATTGACCTTAAAACAATAAATTTTATAATCAGTGTAAAT...ATC'),\n",
       " 'GL000244.1': Seq('GATCTGCGCACCTTGGCCTCCCAAAGTGCTGGGATTACAGGCATGAGCCACCGT...ATC'),\n",
       " 'GL000248.1': Seq('GATCTGTCATTGTCTCCCATGCATGCTTCCCAGCATGCCTGCCTAGCCCCATTC...ATC'),\n",
       " 'GL000196.1': Seq('GATCCCACCAGAGTAGAGGTGGCTATACTAACAACAGAGAAAACAGTCTAAATC...ATC'),\n",
       " 'GL000249.1': Seq('GATCACCAAGGCTGGGAAATATTAGGGAAGGGGGGTTGCCAGACTCCGGACCTG...ATC'),\n",
       " 'GL000246.1': Seq('GATCTTAAGCCTTTGCTGGATGCATCGTTTGCAAATATCTTCTCCCATTGTGTA...ATC'),\n",
       " 'GL000203.1': Seq('GATCCCACAGGATGTGTCTATCTTAAGATTTAGATAAATATTTAAATTTTTTTC...ATC'),\n",
       " 'GL000197.1': Seq('CAAGACTCCTCAGAGTAAGGCATCTGTCCTATTTTGTTCCCCACCCAGGAGTTG...TGG'),\n",
       " 'GL000245.1': Seq('GATCAGATAGGCTTTTGGCTCATTGTAATGAGATGTTCCTATAAACCTAATTTG...ATC'),\n",
       " 'GL000247.1': Seq('GATCTAAGTTTGATTTTTCTGTACAACTCCTTTTCTGAAGTTTTACTCATTTCC...ATC'),\n",
       " 'GL000201.1': Seq('GATCCCACCCAAAGCAGCAGCTTTTTTTTTTTTTTTTTTTTTTTTTGTTATTTT...ATC'),\n",
       " 'GL000235.1': Seq('GATCTCTGAGGTGATGCAACTCTTGTCTAGGCACTGCCTACAGGGGACATTGGT...ATC'),\n",
       " 'GL000239.1': Seq('GATCTCTCTATCGGCCTGCTCTGGGCCACAGAGACTGGCCCTGACGACTTCTCA...ATC'),\n",
       " 'GL000210.1': Seq('AAAAGAACAGCTGGCCTGGGCTTGCTGCTGGGATGCGGGTGGGGCTCTGGGGGG...TCA'),\n",
       " 'GL000231.1': Seq('TTTAATGAATAAGTATTTACGCTATTAATATTTATAATATTGGTGCTATTATTT...ATC'),\n",
       " 'GL000229.1': Seq('GCCGCTGTGGGATTTTCCGTGGGAGAGGTATGGGAGAGGACTGACGCGGCTTCC...ATC'),\n",
       " 'GL000226.1': Seq('GATCTTTCTTTTGATAGAGCAGTTCTGAAAAACACTTTTTGTTGAATCTGCAAG...ATC'),\n",
       " 'GL000207.1': Seq('CATACATTTAATATACCCTCACCATACAGAATGTTCTTTCCCTATTACATAAGG...ATC')}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>color</th>\n",
       "      <th>strand</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>RGB</th>\n",
       "      <th>p</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100008587</td>\n",
       "      <td>100008589</td>\n",
       "      <td>p1@CU680531,0.1352</td>\n",
       "      <td>-89.0</td>\n",
       "      <td>+</td>\n",
       "      <td>100008587</td>\n",
       "      <td>100008589</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p1</td>\n",
       "      <td>CU680531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>p2@LOXL4,0.1291</td>\n",
       "      <td>55.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p2</td>\n",
       "      <td>LOXL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>p3@LOXL4,0.1842</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>30,144,255</td>\n",
       "      <td>p3</td>\n",
       "      <td>LOXL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>p1@LOXL4,0.2200</td>\n",
       "      <td>48.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>30,144,255</td>\n",
       "      <td>p1</td>\n",
       "      <td>LOXL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>p1@PYROXD2,0.2721</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>PYROXD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048119</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9958814</td>\n",
       "      <td>9958815</td>\n",
       "      <td>p@chrY:9958814..9958815,-,0.0551</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>9958814</td>\n",
       "      <td>9958815</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p</td>\n",
       "      <td>chrY:9958814..9958815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048120</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9959226</td>\n",
       "      <td>9959229</td>\n",
       "      <td>p@chrY:9959226..9959229,-,0.0639</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>9959226</td>\n",
       "      <td>9959229</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p</td>\n",
       "      <td>chrY:9959226..9959229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048121</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9968102</td>\n",
       "      <td>9968103</td>\n",
       "      <td>p@chrY:9968102..9968103,-,0.1091</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>9968102</td>\n",
       "      <td>9968103</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p</td>\n",
       "      <td>chrY:9968102..9968103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048122</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9985705</td>\n",
       "      <td>9985711</td>\n",
       "      <td>p@chrY:9985705..9985711,-,0.0853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>9985705</td>\n",
       "      <td>9985711</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p</td>\n",
       "      <td>chrY:9985705..9985711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048123</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9987854</td>\n",
       "      <td>9987855</td>\n",
       "      <td>p@chrY:9987854..9987855,-,0.1054</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>9987854</td>\n",
       "      <td>9987855</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p</td>\n",
       "      <td>chrY:9987854..9987855</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1048124 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome_name      start        end  \\\n",
       "0                 chr10  100008587  100008589   \n",
       "1                 chr10  100015362  100015397   \n",
       "2                 chr10  100017518  100017519   \n",
       "3                 chr10  100027943  100027958   \n",
       "4                 chr10  100174900  100174956   \n",
       "...                 ...        ...        ...   \n",
       "1048119            chrY    9958814    9958815   \n",
       "1048120            chrY    9959226    9959229   \n",
       "1048121            chrY    9968102    9968103   \n",
       "1048122            chrY    9985705    9985711   \n",
       "1048123            chrY    9987854    9987855   \n",
       "\n",
       "                                 locus_id  color strand     start2       end2  \\\n",
       "0                      p1@CU680531,0.1352  -89.0      +  100008587  100008589   \n",
       "1                         p2@LOXL4,0.1291   55.0      -  100015362  100015397   \n",
       "2                         p3@LOXL4,0.1842   13.0      -  100017518  100017519   \n",
       "3                         p1@LOXL4,0.2200   48.0      -  100027943  100027958   \n",
       "4                       p1@PYROXD2,0.2721    0.0      -  100174900  100174956   \n",
       "...                                   ...    ...    ...        ...        ...   \n",
       "1048119  p@chrY:9958814..9958815,-,0.0551    0.0      -    9958814    9958815   \n",
       "1048120  p@chrY:9959226..9959229,-,0.0639    0.0      -    9959226    9959229   \n",
       "1048121  p@chrY:9968102..9968103,-,0.1091    0.0      -    9968102    9968103   \n",
       "1048122  p@chrY:9985705..9985711,-,0.0853    0.0      -    9985705    9985711   \n",
       "1048123  p@chrY:9987854..9987855,-,0.1054    0.0      -    9987854    9987855   \n",
       "\n",
       "                 RGB   p              gene_name  \n",
       "0        211,211,211  p1               CU680531  \n",
       "1        211,211,211  p2                  LOXL4  \n",
       "2         30,144,255  p3                  LOXL4  \n",
       "3         30,144,255  p1                  LOXL4  \n",
       "4         60,179,113  p1                PYROXD2  \n",
       "...              ...  ..                    ...  \n",
       "1048119  211,211,211   p  chrY:9958814..9958815  \n",
       "1048120  211,211,211   p  chrY:9959226..9959229  \n",
       "1048121  211,211,211   p  chrY:9968102..9968103  \n",
       "1048122  211,211,211   p  chrY:9985705..9985711  \n",
       "1048123  211,211,211   p  chrY:9987854..9987855  \n",
       "\n",
       "[1048124 rows x 11 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"..\\\\dataset\\\\TSS_human.bed\", delimiter=\"\\t\", skiprows=1, names=[\"chromosome_name\", \"start\", \"end\", \"locus_id\", \"color\", \"strand\", \"start2\", \"end2\", \"RGB\"])\n",
    "\n",
    "df['p'] = v1(df['locus_id'].values)\n",
    "df['gene_name'] = v2(df['locus_id'].values)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>color</th>\n",
       "      <th>strand</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>RGB</th>\n",
       "      <th>p</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>p2@LOXL4,0.1291</td>\n",
       "      <td>55.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p2</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>p3@LOXL4,0.1842</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>30,144,255</td>\n",
       "      <td>p3</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>p1@LOXL4,0.2200</td>\n",
       "      <td>48.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>30,144,255</td>\n",
       "      <td>p1</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>p1@PYROXD2,0.2721</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>PYROXD2</td>\n",
       "      <td>ENSG00000119943.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174957</td>\n",
       "      <td>100174982</td>\n",
       "      <td>p2@PYROXD2,0.2448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100174957</td>\n",
       "      <td>100174982</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p2</td>\n",
       "      <td>PYROXD2</td>\n",
       "      <td>ENSG00000119943.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254091</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7142206</td>\n",
       "      <td>7142215</td>\n",
       "      <td>p9@PRKY,0.2620</td>\n",
       "      <td>-120.0</td>\n",
       "      <td>+</td>\n",
       "      <td>7142206</td>\n",
       "      <td>7142215</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p9</td>\n",
       "      <td>PRKY</td>\n",
       "      <td>ENSG00000293442.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254092</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7200834</td>\n",
       "      <td>7200838</td>\n",
       "      <td>p12@PRKY,0.0723</td>\n",
       "      <td>-232.0</td>\n",
       "      <td>+</td>\n",
       "      <td>7200834</td>\n",
       "      <td>7200838</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p12</td>\n",
       "      <td>PRKY</td>\n",
       "      <td>ENSG00000099725.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254093</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7200834</td>\n",
       "      <td>7200838</td>\n",
       "      <td>p12@PRKY,0.0723</td>\n",
       "      <td>-232.0</td>\n",
       "      <td>+</td>\n",
       "      <td>7200834</td>\n",
       "      <td>7200838</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p12</td>\n",
       "      <td>PRKY</td>\n",
       "      <td>ENSG00000293442.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254094</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7200913</td>\n",
       "      <td>7200914</td>\n",
       "      <td>p14@PRKY,0.0573</td>\n",
       "      <td>-156.0</td>\n",
       "      <td>+</td>\n",
       "      <td>7200913</td>\n",
       "      <td>7200914</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p14</td>\n",
       "      <td>PRKY</td>\n",
       "      <td>ENSG00000099725.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254095</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7200913</td>\n",
       "      <td>7200914</td>\n",
       "      <td>p14@PRKY,0.0573</td>\n",
       "      <td>-156.0</td>\n",
       "      <td>+</td>\n",
       "      <td>7200913</td>\n",
       "      <td>7200914</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p14</td>\n",
       "      <td>PRKY</td>\n",
       "      <td>ENSG00000293442.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>254096 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome_name      start        end           locus_id  color strand  \\\n",
       "0                chr10  100015362  100015397    p2@LOXL4,0.1291   55.0      -   \n",
       "1                chr10  100017518  100017519    p3@LOXL4,0.1842   13.0      -   \n",
       "2                chr10  100027943  100027958    p1@LOXL4,0.2200   48.0      -   \n",
       "3                chr10  100174900  100174956  p1@PYROXD2,0.2721    0.0      -   \n",
       "4                chr10  100174957  100174982  p2@PYROXD2,0.2448    0.0      -   \n",
       "...                ...        ...        ...                ...    ...    ...   \n",
       "254091            chrY    7142206    7142215     p9@PRKY,0.2620 -120.0      +   \n",
       "254092            chrY    7200834    7200838    p12@PRKY,0.0723 -232.0      +   \n",
       "254093            chrY    7200834    7200838    p12@PRKY,0.0723 -232.0      +   \n",
       "254094            chrY    7200913    7200914    p14@PRKY,0.0573 -156.0      +   \n",
       "254095            chrY    7200913    7200914    p14@PRKY,0.0573 -156.0      +   \n",
       "\n",
       "           start2       end2          RGB    p gene_name             gene_id  \n",
       "0       100015362  100015397  211,211,211   p2     LOXL4   ENSG00000138131.4  \n",
       "1       100017518  100017519   30,144,255   p3     LOXL4   ENSG00000138131.4  \n",
       "2       100027943  100027958   30,144,255   p1     LOXL4   ENSG00000138131.4  \n",
       "3       100174900  100174956   60,179,113   p1   PYROXD2  ENSG00000119943.13  \n",
       "4       100174957  100174982   60,179,113   p2   PYROXD2  ENSG00000119943.13  \n",
       "...           ...        ...          ...  ...       ...                 ...  \n",
       "254091    7142206    7142215   60,179,113   p9      PRKY   ENSG00000293442.1  \n",
       "254092    7200834    7200838  211,211,211  p12      PRKY  ENSG00000099725.15  \n",
       "254093    7200834    7200838  211,211,211  p12      PRKY   ENSG00000293442.1  \n",
       "254094    7200913    7200914  211,211,211  p14      PRKY  ENSG00000099725.15  \n",
       "254095    7200913    7200914  211,211,211  p14      PRKY   ENSG00000293442.1  \n",
       "\n",
       "[254096 rows x 12 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meging=pd.read_csv(\"..\\\\dataset\\\\id_name_world.txt\", sep=\"\\t\")\n",
    "\n",
    "#merging su gene_name\n",
    "df1 = pd.merge(df, meging, on=\"gene_name\")\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>color</th>\n",
       "      <th>strand</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>RGB</th>\n",
       "      <th>p</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>fpkm_median</th>\n",
       "      <th>tpm_median</th>\n",
       "      <th>fpkm_uq_median</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>p2@LOXL4,0.1291</td>\n",
       "      <td>55.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p2</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "      <td>1.800827</td>\n",
       "      <td>3.202559</td>\n",
       "      <td>1.977610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>p3@LOXL4,0.1842</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>30,144,255</td>\n",
       "      <td>p3</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "      <td>1.800827</td>\n",
       "      <td>3.202559</td>\n",
       "      <td>1.977610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>p1@LOXL4,0.2200</td>\n",
       "      <td>48.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>30,144,255</td>\n",
       "      <td>p1</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "      <td>1.800827</td>\n",
       "      <td>3.202559</td>\n",
       "      <td>1.977610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>p1@PYROXD2,0.2721</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>PYROXD2</td>\n",
       "      <td>ENSG00000119943.13</td>\n",
       "      <td>1.167486</td>\n",
       "      <td>2.608927</td>\n",
       "      <td>1.382999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174957</td>\n",
       "      <td>100174982</td>\n",
       "      <td>p2@PYROXD2,0.2448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100174957</td>\n",
       "      <td>100174982</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p2</td>\n",
       "      <td>PYROXD2</td>\n",
       "      <td>ENSG00000119943.13</td>\n",
       "      <td>1.167486</td>\n",
       "      <td>2.608927</td>\n",
       "      <td>1.382999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153925</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2870826</td>\n",
       "      <td>2870844</td>\n",
       "      <td>p2@ZFY-AS1,0.0950</td>\n",
       "      <td>-159.0</td>\n",
       "      <td>-</td>\n",
       "      <td>2870826</td>\n",
       "      <td>2870844</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p2</td>\n",
       "      <td>ZFY-AS1</td>\n",
       "      <td>ENSG00000233070.1</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153926</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778505</td>\n",
       "      <td>6778534</td>\n",
       "      <td>p6@TBL1Y,0.3422</td>\n",
       "      <td>-192.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778505</td>\n",
       "      <td>6778534</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p6</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153927</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778592</td>\n",
       "      <td>6778601</td>\n",
       "      <td>p7@TBL1Y,0.4173</td>\n",
       "      <td>-125.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778592</td>\n",
       "      <td>6778601</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p7</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153928</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778602</td>\n",
       "      <td>6778616</td>\n",
       "      <td>p4@TBL1Y,0.4601</td>\n",
       "      <td>-110.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778602</td>\n",
       "      <td>6778616</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p4</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153929</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778870</td>\n",
       "      <td>6778873</td>\n",
       "      <td>p9@TBL1Y,0.4948</td>\n",
       "      <td>132.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778870</td>\n",
       "      <td>6778873</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p9</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>153930 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome_name      start        end           locus_id  color strand  \\\n",
       "0                chr10  100015362  100015397    p2@LOXL4,0.1291   55.0      -   \n",
       "1                chr10  100017518  100017519    p3@LOXL4,0.1842   13.0      -   \n",
       "2                chr10  100027943  100027958    p1@LOXL4,0.2200   48.0      -   \n",
       "3                chr10  100174900  100174956  p1@PYROXD2,0.2721    0.0      -   \n",
       "4                chr10  100174957  100174982  p2@PYROXD2,0.2448    0.0      -   \n",
       "...                ...        ...        ...                ...    ...    ...   \n",
       "153925            chrY    2870826    2870844  p2@ZFY-AS1,0.0950 -159.0      -   \n",
       "153926            chrY    6778505    6778534    p6@TBL1Y,0.3422 -192.0      +   \n",
       "153927            chrY    6778592    6778601    p7@TBL1Y,0.4173 -125.0      +   \n",
       "153928            chrY    6778602    6778616    p4@TBL1Y,0.4601 -110.0      +   \n",
       "153929            chrY    6778870    6778873    p9@TBL1Y,0.4948  132.0      +   \n",
       "\n",
       "           start2       end2          RGB   p gene_name             gene_id  \\\n",
       "0       100015362  100015397  211,211,211  p2     LOXL4   ENSG00000138131.4   \n",
       "1       100017518  100017519   30,144,255  p3     LOXL4   ENSG00000138131.4   \n",
       "2       100027943  100027958   30,144,255  p1     LOXL4   ENSG00000138131.4   \n",
       "3       100174900  100174956   60,179,113  p1   PYROXD2  ENSG00000119943.13   \n",
       "4       100174957  100174982   60,179,113  p2   PYROXD2  ENSG00000119943.13   \n",
       "...           ...        ...          ...  ..       ...                 ...   \n",
       "153925    2870826    2870844  211,211,211  p2   ZFY-AS1   ENSG00000233070.1   \n",
       "153926    6778505    6778534   60,179,113  p6     TBL1Y  ENSG00000092377.15   \n",
       "153927    6778592    6778601   60,179,113  p7     TBL1Y  ENSG00000092377.15   \n",
       "153928    6778602    6778616   60,179,113  p4     TBL1Y  ENSG00000092377.15   \n",
       "153929    6778870    6778873   60,179,113  p9     TBL1Y  ENSG00000092377.15   \n",
       "\n",
       "        fpkm_median  tpm_median  fpkm_uq_median  \n",
       "0          1.800827    3.202559        1.977610  \n",
       "1          1.800827    3.202559        1.977610  \n",
       "2          1.800827    3.202559        1.977610  \n",
       "3          1.167486    2.608927        1.382999  \n",
       "4          1.167486    2.608927        1.382999  \n",
       "...             ...         ...             ...  \n",
       "153925    -7.643856   -7.643856       -7.643856  \n",
       "153926    -7.643856   -7.643856       -7.643856  \n",
       "153927    -7.643856   -7.643856       -7.643856  \n",
       "153928    -7.643856   -7.643856       -7.643856  \n",
       "153929    -7.643856   -7.643856       -7.643856  \n",
       "\n",
       "[153930 rows x 15 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient=pd.read_csv(\"..\\\\dataset\\\\Dataset_median.csv\", sep=\",\")\n",
    "#merge su gene_id\n",
    "df2 = pd.merge(df1, patient, on=\"gene_id\")\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>color</th>\n",
       "      <th>strand</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>RGB</th>\n",
       "      <th>p</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>fpkm_median</th>\n",
       "      <th>tpm_median</th>\n",
       "      <th>fpkm_uq_median</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>p2@LOXL4,0.1291</td>\n",
       "      <td>55.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100015362</td>\n",
       "      <td>100015397</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p2</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "      <td>1.800827</td>\n",
       "      <td>3.202559</td>\n",
       "      <td>1.977610</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>p3@LOXL4,0.1842</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100017518</td>\n",
       "      <td>100017519</td>\n",
       "      <td>30,144,255</td>\n",
       "      <td>p3</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "      <td>1.800827</td>\n",
       "      <td>3.202559</td>\n",
       "      <td>1.977610</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>p1@LOXL4,0.2200</td>\n",
       "      <td>48.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100027943</td>\n",
       "      <td>100027958</td>\n",
       "      <td>30,144,255</td>\n",
       "      <td>p1</td>\n",
       "      <td>LOXL4</td>\n",
       "      <td>ENSG00000138131.4</td>\n",
       "      <td>1.800827</td>\n",
       "      <td>3.202559</td>\n",
       "      <td>1.977610</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>p1@PYROXD2,0.2721</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100174900</td>\n",
       "      <td>100174956</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>PYROXD2</td>\n",
       "      <td>ENSG00000119943.13</td>\n",
       "      <td>1.167486</td>\n",
       "      <td>2.608927</td>\n",
       "      <td>1.382999</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100174957</td>\n",
       "      <td>100174982</td>\n",
       "      <td>p2@PYROXD2,0.2448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>100174957</td>\n",
       "      <td>100174982</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p2</td>\n",
       "      <td>PYROXD2</td>\n",
       "      <td>ENSG00000119943.13</td>\n",
       "      <td>1.167486</td>\n",
       "      <td>2.608927</td>\n",
       "      <td>1.382999</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153925</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2870826</td>\n",
       "      <td>2870844</td>\n",
       "      <td>p2@ZFY-AS1,0.0950</td>\n",
       "      <td>-159.0</td>\n",
       "      <td>-</td>\n",
       "      <td>2870826</td>\n",
       "      <td>2870844</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p2</td>\n",
       "      <td>ZFY-AS1</td>\n",
       "      <td>ENSG00000233070.1</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153926</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778505</td>\n",
       "      <td>6778534</td>\n",
       "      <td>p6@TBL1Y,0.3422</td>\n",
       "      <td>-192.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778505</td>\n",
       "      <td>6778534</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p6</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153927</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778592</td>\n",
       "      <td>6778601</td>\n",
       "      <td>p7@TBL1Y,0.4173</td>\n",
       "      <td>-125.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778592</td>\n",
       "      <td>6778601</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p7</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153928</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778602</td>\n",
       "      <td>6778616</td>\n",
       "      <td>p4@TBL1Y,0.4601</td>\n",
       "      <td>-110.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778602</td>\n",
       "      <td>6778616</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p4</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153929</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6778870</td>\n",
       "      <td>6778873</td>\n",
       "      <td>p9@TBL1Y,0.4948</td>\n",
       "      <td>132.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6778870</td>\n",
       "      <td>6778873</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p9</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>ENSG00000092377.15</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>153930 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome_name      start        end           locus_id  color strand  \\\n",
       "0                chr10  100015362  100015397    p2@LOXL4,0.1291   55.0      -   \n",
       "1                chr10  100017518  100017519    p3@LOXL4,0.1842   13.0      -   \n",
       "2                chr10  100027943  100027958    p1@LOXL4,0.2200   48.0      -   \n",
       "3                chr10  100174900  100174956  p1@PYROXD2,0.2721    0.0      -   \n",
       "4                chr10  100174957  100174982  p2@PYROXD2,0.2448    0.0      -   \n",
       "...                ...        ...        ...                ...    ...    ...   \n",
       "153925            chrY    2870826    2870844  p2@ZFY-AS1,0.0950 -159.0      -   \n",
       "153926            chrY    6778505    6778534    p6@TBL1Y,0.3422 -192.0      +   \n",
       "153927            chrY    6778592    6778601    p7@TBL1Y,0.4173 -125.0      +   \n",
       "153928            chrY    6778602    6778616    p4@TBL1Y,0.4601 -110.0      +   \n",
       "153929            chrY    6778870    6778873    p9@TBL1Y,0.4948  132.0      +   \n",
       "\n",
       "           start2       end2          RGB   p gene_name             gene_id  \\\n",
       "0       100015362  100015397  211,211,211  p2     LOXL4   ENSG00000138131.4   \n",
       "1       100017518  100017519   30,144,255  p3     LOXL4   ENSG00000138131.4   \n",
       "2       100027943  100027958   30,144,255  p1     LOXL4   ENSG00000138131.4   \n",
       "3       100174900  100174956   60,179,113  p1   PYROXD2  ENSG00000119943.13   \n",
       "4       100174957  100174982   60,179,113  p2   PYROXD2  ENSG00000119943.13   \n",
       "...           ...        ...          ...  ..       ...                 ...   \n",
       "153925    2870826    2870844  211,211,211  p2   ZFY-AS1   ENSG00000233070.1   \n",
       "153926    6778505    6778534   60,179,113  p6     TBL1Y  ENSG00000092377.15   \n",
       "153927    6778592    6778601   60,179,113  p7     TBL1Y  ENSG00000092377.15   \n",
       "153928    6778602    6778616   60,179,113  p4     TBL1Y  ENSG00000092377.15   \n",
       "153929    6778870    6778873   60,179,113  p9     TBL1Y  ENSG00000092377.15   \n",
       "\n",
       "        fpkm_median  tpm_median  fpkm_uq_median  p_value  \n",
       "0          1.800827    3.202559        1.977610        2  \n",
       "1          1.800827    3.202559        1.977610        3  \n",
       "2          1.800827    3.202559        1.977610        1  \n",
       "3          1.167486    2.608927        1.382999        1  \n",
       "4          1.167486    2.608927        1.382999        2  \n",
       "...             ...         ...             ...      ...  \n",
       "153925    -7.643856   -7.643856       -7.643856        2  \n",
       "153926    -7.643856   -7.643856       -7.643856        6  \n",
       "153927    -7.643856   -7.643856       -7.643856        7  \n",
       "153928    -7.643856   -7.643856       -7.643856        4  \n",
       "153929    -7.643856   -7.643856       -7.643856        9  \n",
       "\n",
       "[153930 rows x 16 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "def estrai_numero(s):\n",
    "    match = re.match(r'p(\\d+)', s)\n",
    "    if match:\n",
    "        return int(match.group(1))\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Applica la funzione alla colonna del DataFrame\n",
    "df2['p_value'] = df2['p'].apply(estrai_numero)\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "p\n",
       "p1    14075\n",
       "p2       86\n",
       "p3       14\n",
       "p4       10\n",
       "p5        5\n",
       "p6        3\n",
       "p7        2\n",
       "p8        2\n",
       "p9        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df3 = df2.loc[df2.groupby(\"gene_name\")[\"p_value\"].idxmin()]\n",
    "\n",
    "\n",
    "\n",
    "df3.value_counts(\"p\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>color</th>\n",
       "      <th>strand</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>RGB</th>\n",
       "      <th>p</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>fpkm_median</th>\n",
       "      <th>tpm_median</th>\n",
       "      <th>fpkm_uq_median</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>153866</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99839925</td>\n",
       "      <td>99839938</td>\n",
       "      <td>p1@TNMD,0.1353</td>\n",
       "      <td>126.0</td>\n",
       "      <td>+</td>\n",
       "      <td>99839925</td>\n",
       "      <td>99839938</td>\n",
       "      <td>211,211,211</td>\n",
       "      <td>p1</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>-3.925769</td>\n",
       "      <td>-2.475535</td>\n",
       "      <td>-3.733124</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38903</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169863016</td>\n",
       "      <td>169863109</td>\n",
       "      <td>p1@SCYL3,0.2639</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>169863016</td>\n",
       "      <td>169863109</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>0.873892</td>\n",
       "      <td>2.396543</td>\n",
       "      <td>1.063089</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42198</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27961629</td>\n",
       "      <td>27961665</td>\n",
       "      <td>p1@FGR,0.5037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>27961629</td>\n",
       "      <td>27961665</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>FGR</td>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "      <td>0.612022</td>\n",
       "      <td>2.081715</td>\n",
       "      <td>0.757621</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71796</th>\n",
       "      <td>chr6</td>\n",
       "      <td>143832820</td>\n",
       "      <td>143832875</td>\n",
       "      <td>p1@FUCA2,0.4000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>143832820</td>\n",
       "      <td>143832875</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>FUCA2</td>\n",
       "      <td>ENSG00000001036.14</td>\n",
       "      <td>3.827545</td>\n",
       "      <td>5.333456</td>\n",
       "      <td>4.001946</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74636</th>\n",
       "      <td>chr6</td>\n",
       "      <td>53409890</td>\n",
       "      <td>53409906</td>\n",
       "      <td>p1@GCLC,0.4228</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>-</td>\n",
       "      <td>53409890</td>\n",
       "      <td>53409906</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>GCLC</td>\n",
       "      <td>ENSG00000001084.13</td>\n",
       "      <td>1.509544</td>\n",
       "      <td>2.956224</td>\n",
       "      <td>1.694702</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86592</th>\n",
       "      <td>chrX</td>\n",
       "      <td>101967366</td>\n",
       "      <td>101967391</td>\n",
       "      <td>p1@ARMCX5-GPRASP2,p1@GPRASP2,0.6351</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>101967366</td>\n",
       "      <td>101967391</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>ARMCX5-GPRASP2</td>\n",
       "      <td>ENSG00000286237.1</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65961</th>\n",
       "      <td>chr4</td>\n",
       "      <td>89513578</td>\n",
       "      <td>89513670</td>\n",
       "      <td>p1@HERC3,0.3732</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>89513578</td>\n",
       "      <td>89513670</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>HERC3</td>\n",
       "      <td>ENSG00000287542.1</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82689</th>\n",
       "      <td>chr8</td>\n",
       "      <td>67579807</td>\n",
       "      <td>67579860</td>\n",
       "      <td>p1@C8orf44,p1@SGK3,0.4647</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>67579807</td>\n",
       "      <td>67579860</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>C8orf44</td>\n",
       "      <td>ENSG00000288596.2</td>\n",
       "      <td>0.995448</td>\n",
       "      <td>2.458933</td>\n",
       "      <td>1.200316</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148900</th>\n",
       "      <td>chr8</td>\n",
       "      <td>53852965</td>\n",
       "      <td>53852990</td>\n",
       "      <td>p1@NPBWR1,0.2996</td>\n",
       "      <td>497.0</td>\n",
       "      <td>+</td>\n",
       "      <td>53852965</td>\n",
       "      <td>53852990</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>NPBWR1</td>\n",
       "      <td>ENSG00000288611.1</td>\n",
       "      <td>-6.687800</td>\n",
       "      <td>-5.779918</td>\n",
       "      <td>-6.546245</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87371</th>\n",
       "      <td>chrX</td>\n",
       "      <td>139866699</td>\n",
       "      <td>139866741</td>\n",
       "      <td>p1@CDR1,0.6297</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>139866699</td>\n",
       "      <td>139866741</td>\n",
       "      <td>60,179,113</td>\n",
       "      <td>p1</td>\n",
       "      <td>CDR1</td>\n",
       "      <td>ENSG00000288642.1</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>-7.643856</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14272 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome_name      start        end  \\\n",
       "153866            chrX   99839925   99839938   \n",
       "38903             chr1  169863016  169863109   \n",
       "42198             chr1   27961629   27961665   \n",
       "71796             chr6  143832820  143832875   \n",
       "74636             chr6   53409890   53409906   \n",
       "...                ...        ...        ...   \n",
       "86592             chrX  101967366  101967391   \n",
       "65961             chr4   89513578   89513670   \n",
       "82689             chr8   67579807   67579860   \n",
       "148900            chr8   53852965   53852990   \n",
       "87371             chrX  139866699  139866741   \n",
       "\n",
       "                                   locus_id  color strand     start2  \\\n",
       "153866                       p1@TNMD,0.1353  126.0      +   99839925   \n",
       "38903                       p1@SCYL3,0.2639    0.0      -  169863016   \n",
       "42198                         p1@FGR,0.5037    0.0      -   27961629   \n",
       "71796                       p1@FUCA2,0.4000    0.0      -  143832820   \n",
       "74636                        p1@GCLC,0.4228   -4.0      -   53409890   \n",
       "...                                     ...    ...    ...        ...   \n",
       "86592   p1@ARMCX5-GPRASP2,p1@GPRASP2,0.6351    0.0      +  101967366   \n",
       "65961                       p1@HERC3,0.3732    0.0      +   89513578   \n",
       "82689             p1@C8orf44,p1@SGK3,0.4647    0.0      +   67579807   \n",
       "148900                     p1@NPBWR1,0.2996  497.0      +   53852965   \n",
       "87371                        p1@CDR1,0.6297    0.0      -  139866699   \n",
       "\n",
       "             end2          RGB   p       gene_name             gene_id  \\\n",
       "153866   99839938  211,211,211  p1            TNMD   ENSG00000000005.6   \n",
       "38903   169863109   60,179,113  p1           SCYL3  ENSG00000000457.14   \n",
       "42198    27961665   60,179,113  p1             FGR  ENSG00000000938.13   \n",
       "71796   143832875   60,179,113  p1           FUCA2  ENSG00000001036.14   \n",
       "74636    53409906   60,179,113  p1            GCLC  ENSG00000001084.13   \n",
       "...           ...          ...  ..             ...                 ...   \n",
       "86592   101967391   60,179,113  p1  ARMCX5-GPRASP2   ENSG00000286237.1   \n",
       "65961    89513670   60,179,113  p1           HERC3   ENSG00000287542.1   \n",
       "82689    67579860   60,179,113  p1         C8orf44   ENSG00000288596.2   \n",
       "148900   53852990   60,179,113  p1          NPBWR1   ENSG00000288611.1   \n",
       "87371   139866741   60,179,113  p1            CDR1   ENSG00000288642.1   \n",
       "\n",
       "        fpkm_median  tpm_median  fpkm_uq_median  p_value  \n",
       "153866    -3.925769   -2.475535       -3.733124        1  \n",
       "38903      0.873892    2.396543        1.063089        1  \n",
       "42198      0.612022    2.081715        0.757621        1  \n",
       "71796      3.827545    5.333456        4.001946        1  \n",
       "74636      1.509544    2.956224        1.694702        1  \n",
       "...             ...         ...             ...      ...  \n",
       "86592     -7.643856   -7.643856       -7.643856        1  \n",
       "65961     -7.643856   -7.643856       -7.643856        1  \n",
       "82689      0.995448    2.458933        1.200316        1  \n",
       "148900    -6.687800   -5.779918       -6.546245        1  \n",
       "87371     -7.643856   -7.643856       -7.643856        1  \n",
       "\n",
       "[14272 rows x 16 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df4 = df2.loc[df2.groupby(\"gene_id\")[\"p_value\"].idxmin()]\n",
    "df4"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "AIE",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
